{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ae6e2cb-8406-4873-96df-893bee13ace3",
   "metadata": {},
   "source": [
    "run with `popstrucure` env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "101a4f4b-10b0-4add-bad8-2effdecd0c72",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Unable to determine R home: [Errno 2] No such file or directory: 'R'\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import vcf\n",
    "from pathlib import Path\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import cm\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "import glob\n",
    "import numpy as np\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as mlines\n",
    "import statistics\n",
    "import allel\n",
    "from collections import defaultdict\n",
    "from itertools import combinations\n",
    "from haversine import haversine\n",
    "from skbio.stats.distance import mantel\n",
    "from scipy.stats import ttest_ind\n",
    "import cartopy.crs as ccrs\n",
    "import cartopy.feature as cf\n",
    "from pykrige.ok import OrdinaryKriging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fa2e0624-d18c-44e8-af03-8217a93dbcd2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_scan\"\n",
    "results_dir=\"{}/results\".format(proj_dir)\n",
    "\n",
    "os.chdir(proj_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4fc670a7-a613-4d36-b9d8-a130caf75cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_df=pd.read_csv(\"{}/data/seq_and_sample_docs/all_sh_sb_sample_data.csv\".format(proj_dir), sep=\",\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99e8dc1c-72ed-4fd3-bc88-650ef81e6f2e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Genome Coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c18daabd-2a4b-4cd9-af83-25e19292d294",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"{}/genome_coverage\".format(results_dir)).mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(\"{}/genome_coverage\".format(results_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0a09c14b-9736-453b-887d-ac0043a529d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_df=pd.read_csv(\"../initial_pca/pca_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4ef62cc6-7f8b-47a4-8492-db3fde772960",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>locale</th>\n",
       "      <th>host</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>project</th>\n",
       "      <th>data_type</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>kmeans_group</th>\n",
       "      <th>pca_species</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>-0.174693</td>\n",
       "      <td>0.025622</td>\n",
       "      <td>0.001506</td>\n",
       "      <td>0.004149</td>\n",
       "      <td>0.035574</td>\n",
       "      <td>0.055396</td>\n",
       "      <td>0.032718</td>\n",
       "      <td>-0.019520</td>\n",
       "      <td>0.007569</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>-0.173848</td>\n",
       "      <td>0.025251</td>\n",
       "      <td>0.001886</td>\n",
       "      <td>0.005161</td>\n",
       "      <td>0.036499</td>\n",
       "      <td>0.060013</td>\n",
       "      <td>0.032135</td>\n",
       "      <td>-0.019021</td>\n",
       "      <td>0.003526</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>-0.174493</td>\n",
       "      <td>0.025294</td>\n",
       "      <td>0.001699</td>\n",
       "      <td>0.001562</td>\n",
       "      <td>0.037332</td>\n",
       "      <td>0.060250</td>\n",
       "      <td>0.030712</td>\n",
       "      <td>-0.022199</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>-0.173862</td>\n",
       "      <td>0.025171</td>\n",
       "      <td>0.003061</td>\n",
       "      <td>-0.001638</td>\n",
       "      <td>0.037943</td>\n",
       "      <td>0.059797</td>\n",
       "      <td>0.031391</td>\n",
       "      <td>-0.015780</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>-0.173011</td>\n",
       "      <td>0.025618</td>\n",
       "      <td>0.001527</td>\n",
       "      <td>0.005455</td>\n",
       "      <td>0.038614</td>\n",
       "      <td>0.061741</td>\n",
       "      <td>0.032516</td>\n",
       "      <td>-0.022343</td>\n",
       "      <td>0.009809</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.023482</td>\n",
       "      <td>-0.078226</td>\n",
       "      <td>-0.011342</td>\n",
       "      <td>-0.059942</td>\n",
       "      <td>0.017520</td>\n",
       "      <td>-0.043266</td>\n",
       "      <td>0.041449</td>\n",
       "      <td>-0.016997</td>\n",
       "      <td>0.021684</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.022237</td>\n",
       "      <td>-0.077112</td>\n",
       "      <td>-0.019412</td>\n",
       "      <td>-0.059978</td>\n",
       "      <td>0.006374</td>\n",
       "      <td>-0.004766</td>\n",
       "      <td>0.030555</td>\n",
       "      <td>-0.041763</td>\n",
       "      <td>0.080463</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.023857</td>\n",
       "      <td>-0.073894</td>\n",
       "      <td>-0.020205</td>\n",
       "      <td>-0.050080</td>\n",
       "      <td>0.015119</td>\n",
       "      <td>-0.015182</td>\n",
       "      <td>0.038133</td>\n",
       "      <td>-0.019263</td>\n",
       "      <td>0.036911</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>0.026630</td>\n",
       "      <td>-0.083945</td>\n",
       "      <td>-0.003698</td>\n",
       "      <td>-0.069531</td>\n",
       "      <td>-0.009745</td>\n",
       "      <td>-0.085123</td>\n",
       "      <td>0.095988</td>\n",
       "      <td>-0.108937</td>\n",
       "      <td>0.082528</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>0.053261</td>\n",
       "      <td>0.112131</td>\n",
       "      <td>-0.096936</td>\n",
       "      <td>-0.011639</td>\n",
       "      <td>-0.039766</td>\n",
       "      <td>0.015451</td>\n",
       "      <td>0.005055</td>\n",
       "      <td>-0.002872</td>\n",
       "      <td>-0.008524</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-15.783442</td>\n",
       "      <td>28.172582</td>\n",
       "      <td>1</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_se</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample_name         1         2         3         4         5  \\\n",
       "0            Sb_NG_au_1.2 -0.174693  0.025622  0.001506  0.004149  0.035574   \n",
       "1           Sb_NG_au_2.10 -0.173848  0.025251  0.001886  0.005161  0.036499   \n",
       "2           Sb_NG_au_2.13 -0.174493  0.025294  0.001699  0.001562  0.037332   \n",
       "3            Sb_NG_au_2.5 -0.173862  0.025171  0.003061 -0.001638  0.037943   \n",
       "4            Sb_NG_au_2.6 -0.173011  0.025618  0.001527  0.005455  0.038614   \n",
       "..                    ...       ...       ...       ...       ...       ...   \n",
       "166  ssp_niger_libore_167  0.023482 -0.078226 -0.011342 -0.059942  0.017520   \n",
       "167  ssp_niger_libore_168  0.022237 -0.077112 -0.019412 -0.059978  0.006374   \n",
       "168  ssp_niger_libore_169  0.023857 -0.073894 -0.020205 -0.050080  0.015119   \n",
       "169  ssp_niger_libore_173  0.026630 -0.083945 -0.003698 -0.069531 -0.009745   \n",
       "170   ssp_zambia_kafue_71  0.053261  0.112131 -0.096936 -0.011639 -0.039766   \n",
       "\n",
       "            6         7         8         9  ...  locale   host  phenotype  \\\n",
       "0    0.055396  0.032718 -0.019520  0.007569  ...      na    cow      bovis   \n",
       "1    0.060013  0.032135 -0.019021  0.003526  ...      na    cow      bovis   \n",
       "2    0.060250  0.030712 -0.022199  0.003399  ...      na    cow      bovis   \n",
       "3    0.059797  0.031391 -0.015780  0.000577  ...      na    cow      bovis   \n",
       "4    0.061741  0.032516 -0.022343  0.009809  ...      na    cow      bovis   \n",
       "..        ...       ...       ...       ...  ...     ...    ...        ...   \n",
       "166 -0.043266  0.041449 -0.016997  0.021684  ...      na  snail        unk   \n",
       "167 -0.004766  0.030555 -0.041763  0.080463  ...      na  snail        unk   \n",
       "168 -0.015182  0.038133 -0.019263  0.036911  ...      na  snail        unk   \n",
       "169 -0.085123  0.095988 -0.108937  0.082528  ...      na  snail        unk   \n",
       "170  0.015451  0.005055 -0.002872 -0.008524  ...      na  snail        unk   \n",
       "\n",
       "         project  data_type        lat        lon  kmeans_group  \\\n",
       "0        nigeria     genome   7.067448   6.271139             2   \n",
       "1        nigeria     genome   7.067448   6.271139             2   \n",
       "2        nigeria     genome   7.067448   6.271139             2   \n",
       "3        nigeria     genome   7.067448   6.271139             2   \n",
       "4        nigeria     genome   7.067448   6.271139             2   \n",
       "..           ...        ...        ...        ...           ...   \n",
       "166  scan_africa     genome  13.390080   2.197110             0   \n",
       "167  scan_africa     genome  13.390080   2.197110             0   \n",
       "168  scan_africa     genome  13.404170   2.188020             0   \n",
       "169  scan_africa     genome  13.404170   2.188020             0   \n",
       "170  scan_africa     genome -15.783442  28.172582             1   \n",
       "\n",
       "       pca_species  kmeans_label  \n",
       "0          s_bovis            sb  \n",
       "1          s_bovis            sb  \n",
       "2          s_bovis            sb  \n",
       "3          s_bovis            sb  \n",
       "4          s_bovis            sb  \n",
       "..             ...           ...  \n",
       "166  s_haematobium         sh_nw  \n",
       "167  s_haematobium         sh_nw  \n",
       "168  s_haematobium         sh_nw  \n",
       "169  s_haematobium         sh_nw  \n",
       "170  s_haematobium         sh_se  \n",
       "\n",
       "[171 rows x 37 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2d1bbee4-b73a-419e-92cd-670da09d26db",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pca_df[~pca_df['sample_name'].isin(public_samples)][\"sample_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc0b45fe-6b1a-4f37-8216-6f4a8efc7788",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples.to_csv(\"samples.list\", header=None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "f79613c9-2595-47da-92b8-b2fdf931b206",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your job 813290 (\"Sb_NG_au_1.2\") has been submitted\n",
      "Your job 813291 (\"Sb_NG_au_2.10\") has been submitted\n",
      "Your job 813292 (\"Sb_NG_au_2.13\") has been submitted\n",
      "Your job 813293 (\"Sb_NG_au_2.5\") has been submitted\n",
      "Your job 813294 (\"Sb_NG_au_2.6\") has been submitted\n",
      "Your job 813295 (\"Sb_NG_be_1.10\") has been submitted\n",
      "Your job 813296 (\"Sb_NG_be_1.3\") has been submitted\n",
      "Your job 813297 (\"Sb_NG_be_1.5\") has been submitted\n",
      "Your job 813298 (\"Sb_NG_en_1.1\") has been submitted\n",
      "Your job 813299 (\"Sh_NG_eb_6_2\") has been submitted\n",
      "Your job 813300 (\"Sh_NG_ed_1_3\") has been submitted\n",
      "Your job 813301 (\"Sh_NG_ed_3_2\") has been submitted\n",
      "Your job 813302 (\"Sh_NG_kb_2_1\") has been submitted\n",
      "Your job 813303 (\"Sh_NG_kw_1_10\") has been submitted\n",
      "Your job 813304 (\"Sh_NG_os_1_4\") has been submitted\n",
      "Your job 813305 (\"sbo_cdivoire_foro_106\") has been submitted\n",
      "Your job 813306 (\"sbo_cdivoire_foro_116\") has been submitted\n",
      "Your job 813307 (\"sbo_cdivoire_noumousso_107\") has been submitted\n",
      "Your job 813308 (\"sbo_cdivoire_noumousso_117\") has been submitted\n",
      "Your job 813309 (\"sbo_cdivoire_raviart_108\") has been submitted\n",
      "Your job 813310 (\"sbo_cdivoire_raviart_109\") has been submitted\n",
      "Your job 813311 (\"sbo_niger_libore_154\") has been submitted\n",
      "Your job 813312 (\"sbo_niger_libore_155\") has been submitted\n",
      "Your job 813313 (\"sbo_niger_libore_160\") has been submitted\n",
      "Your job 813314 (\"sbo_niger_libore_162\") has been submitted\n",
      "Your job 813315 (\"sbo_niger_libore_163\") has been submitted\n",
      "Your job 813316 (\"sbo_niger_libore_164\") has been submitted\n",
      "Your job 813317 (\"sbo_niger_libore_170\") has been submitted\n",
      "Your job 813318 (\"sbo_niger_libore_171\") has been submitted\n",
      "Your job 813319 (\"sbo_niger_libore_172\") has been submitted\n",
      "Your job 813320 (\"sbo_niger_libore_175\") has been submitted\n",
      "Your job 813321 (\"sbo_niger_libore_176\") has been submitted\n",
      "Your job 813322 (\"sbo_uganda_apac13_35\") has been submitted\n",
      "Your job 813323 (\"sbo_uganda_runga_44\") has been submitted\n",
      "Your job 813324 (\"sh_cameroon_SRR13579883\") has been submitted\n",
      "Your job 813325 (\"sh_egypt_SRR13579865\") has been submitted\n",
      "Your job 813326 (\"sh_egypt_SRR433862\") has been submitted\n",
      "Your job 813327 (\"sh_gambia_SRR13579872\") has been submitted\n",
      "Your job 813328 (\"sh_guineabissau_SRR13579870\") has been submitted\n",
      "Your job 813329 (\"sh_ivorycoast_SRR13579866\") has been submitted\n",
      "Your job 813330 (\"sh_mali_SRR13579867\") has been submitted\n",
      "Your job 813331 (\"sh_niger_SRR8284786\") has been submitted\n",
      "Your job 813332 (\"sh_niger_SRR8284787\") has been submitted\n",
      "Your job 813333 (\"sh_niger_SRR8284788\") has been submitted\n",
      "Your job 813334 (\"sh_niger_SRR8284789\") has been submitted\n",
      "Your job 813335 (\"sh_niger_SRR8284790\") has been submitted\n",
      "Your job 813336 (\"sh_niger_SRR8284791\") has been submitted\n",
      "Your job 813337 (\"sh_nigeria_SRR13579869\") has been submitted\n",
      "Your job 813338 (\"sh_senegal_SRR13579881\") has been submitted\n",
      "Your job 813339 (\"sh_tzpem_SRR8284792\") has been submitted\n",
      "Your job 813340 (\"sh_tzpem_SRR8284794\") has been submitted\n",
      "Your job 813341 (\"sh_tzpem_SRR8284796\") has been submitted\n",
      "Your job 813342 (\"sh_tzpem_SRR8284797\") has been submitted\n",
      "Your job 813343 (\"sh_zanzibar_SRR13579868\") has been submitted\n",
      "Your job 813344 (\"sha_angola_cota_57\") has been submitted\n",
      "Your job 813345 (\"sha_angola_cota_58\") has been submitted\n",
      "Your job 813346 (\"sha_angola_cota_59\") has been submitted\n",
      "Your job 813347 (\"sha_angola_icau_60\") has been submitted\n",
      "Your job 813348 (\"sha_angola_icau_61\") has been submitted\n",
      "Your job 813349 (\"sha_angola_icau_62\") has been submitted\n",
      "Your job 813350 (\"sha_angola_icau_63\") has been submitted\n",
      "Your job 813351 (\"sha_angola_icau_64\") has been submitted\n",
      "Your job 813352 (\"sha_cdivoire_kongobo_115\") has been submitted\n",
      "Your job 813353 (\"sha_kenya_unk_49\") has been submitted\n",
      "Your job 813354 (\"sha_liberia_dormeyan_54\") has been submitted\n",
      "Your job 813355 (\"sha_liberia_joekpenmue_55\") has been submitted\n",
      "Your job 813356 (\"sha_madag_belesalampy_73\") has been submitted\n",
      "Your job 813357 (\"sha_madag_belesalampy_74\") has been submitted\n",
      "Your job 813358 (\"sha_madag_belesalampy_75\") has been submitted\n",
      "Your job 813359 (\"sha_madag_belesalampy_76\") has been submitted\n",
      "Your job 813360 (\"sha_madag_belesalampy_77\") has been submitted\n",
      "Your job 813361 (\"sha_madag_belesalampy_78\") has been submitted\n",
      "Your job 813362 (\"sha_madag_belesalampy_80\") has been submitted\n",
      "Your job 813363 (\"sha_madag_belesalampy_81\") has been submitted\n",
      "Your job 813364 (\"sha_madag_belesalampy_82\") has been submitted\n",
      "Your job 813365 (\"sha_madag_belesalampy_83\") has been submitted\n",
      "Your job 813366 (\"sha_madag_belesalampy_84\") has been submitted\n",
      "Your job 813367 (\"sha_madag_belesalampy_87\") has been submitted\n",
      "Your job 813368 (\"sha_namib_mayenzere_134\") has been submitted\n",
      "Your job 813369 (\"sha_niger_libore_138\") has been submitted\n",
      "Your job 813370 (\"sha_niger_libore_139\") has been submitted\n",
      "Your job 813371 (\"sha_niger_libore_140\") has been submitted\n",
      "Your job 813372 (\"sha_niger_libore_141\") has been submitted\n",
      "Your job 813373 (\"sha_niger_libore_142\") has been submitted\n",
      "Your job 813374 (\"sha_niger_libore_143\") has been submitted\n",
      "Your job 813375 (\"sha_niger_libore_146\") has been submitted\n",
      "Your job 813376 (\"sha_niger_libore_147\") has been submitted\n",
      "Your job 813377 (\"sha_niger_libore_148\") has been submitted\n",
      "Your job 813378 (\"sha_niger_libore_149\") has been submitted\n",
      "Your job 813379 (\"sha_niger_libore_150\") has been submitted\n",
      "Your job 813380 (\"sha_niger_libore_151\") has been submitted\n",
      "Your job 813381 (\"sha_pemba_uwandani_1\") has been submitted\n",
      "Your job 813382 (\"sha_pemba_uwandani_10\") has been submitted\n",
      "Your job 813383 (\"sha_pemba_uwandani_11\") has been submitted\n",
      "Your job 813384 (\"sha_pemba_uwandani_12\") has been submitted\n",
      "Your job 813385 (\"sha_pemba_uwandani_13\") has been submitted\n",
      "Your job 813386 (\"sha_pemba_uwandani_15\") has been submitted\n",
      "Your job 813387 (\"sha_pemba_uwandani_2\") has been submitted\n",
      "Your job 813388 (\"sha_pemba_uwandani_5\") has been submitted\n",
      "Your job 813389 (\"sha_pemba_uwandani_7\") has been submitted\n",
      "Your job 813390 (\"sha_sudan_canal4_125\") has been submitted\n",
      "Your job 813391 (\"sha_sudan_canal4_126\") has been submitted\n",
      "Your job 813392 (\"sha_sudan_canal4_127\") has been submitted\n",
      "Your job 813393 (\"sha_sudan_canal4_132\") has been submitted\n",
      "Your job 813394 (\"sha_sudan_schoole_120\") has been submitted\n",
      "Your job 813395 (\"sha_sudan_schoolh_123\") has been submitted\n",
      "Your job 813396 (\"sha_sudan_schooli_129\") has been submitted\n",
      "Your job 813397 (\"sha_sudan_schooli_130\") has been submitted\n",
      "Your job 813398 (\"sha_swaz_mkhuzweni_177\") has been submitted\n",
      "Your job 813399 (\"sha_swaz_mkhuzweni_178\") has been submitted\n",
      "Your job 813400 (\"sha_swaz_mkhuzweni_179\") has been submitted\n",
      "Your job 813401 (\"sha_swaz_mkhuzweni_180\") has been submitted\n",
      "Your job 813402 (\"sha_swaz_mkhuzweni_182\") has been submitted\n",
      "Your job 813403 (\"sha_swaz_mkhuzweni_184\") has been submitted\n",
      "Your job 813404 (\"sha_swaz_njojane_188\") has been submitted\n",
      "Your job 813405 (\"sha_swaz_njojane_190\") has been submitted\n",
      "Your job 813406 (\"sha_swaz_njojane_191\") has been submitted\n",
      "Your job 813407 (\"sha_swaz_qomintaba_185\") has been submitted\n",
      "Your job 813408 (\"sha_swaz_unk_186\") has been submitted\n",
      "Your job 813409 (\"sha_swaz_unk_187\") has been submitted\n",
      "Your job 813410 (\"sha_unguja_kinyasini_16\") has been submitted\n",
      "Your job 813411 (\"sha_unguja_kinyasini_19\") has been submitted\n",
      "Your job 813412 (\"sha_unguja_kinyasini_23\") has been submitted\n",
      "Your job 813413 (\"sha_unguja_kinyasini_25\") has been submitted\n",
      "Your job 813414 (\"sha_unguja_kinyasini_26\") has been submitted\n",
      "Your job 813415 (\"sha_unguja_kinyasini_27\") has been submitted\n",
      "Your job 813416 (\"sha_unguja_kinyasini_28\") has been submitted\n",
      "Your job 813417 (\"sha_unguja_kinyasini_29\") has been submitted\n",
      "Your job 813418 (\"sha_unguja_kinyasini_30\") has been submitted\n",
      "Your job 813419 (\"sha_zambia_lishiko_66\") has been submitted\n",
      "Your job 813420 (\"ssp_cdivoire_allokokro_102\") has been submitted\n",
      "Your job 813421 (\"ssp_cdivoire_allokokro_104\") has been submitted\n",
      "Your job 813422 (\"ssp_cdivoire_foro_101\") has been submitted\n",
      "Your job 813423 (\"ssp_cdivoire_foro_103\") has been submitted\n",
      "Your job 813424 (\"ssp_cdivoire_kongobo_95\") has been submitted\n",
      "Your job 813425 (\"ssp_cdivoire_kongobo_96\") has been submitted\n",
      "Your job 813426 (\"ssp_cdivoire_kongobo_97\") has been submitted\n",
      "Your job 813427 (\"ssp_cdivoire_kongobo_98\") has been submitted\n",
      "Your job 813428 (\"ssp_cdivoire_linguebo_100\") has been submitted\n",
      "Your job 813429 (\"ssp_cdivoire_linguebo_105\") has been submitted\n",
      "Your job 813430 (\"ssp_cdivoire_noumousso_88\") has been submitted\n",
      "Your job 813431 (\"ssp_cdivoire_noumousso_94\") has been submitted\n",
      "Your job 813432 (\"ssp_cdivoire_raviart_110\") has been submitted\n",
      "Your job 813433 (\"ssp_cdivoire_raviart_111\") has been submitted\n",
      "Your job 813434 (\"ssp_cdivoire_raviart_113\") has been submitted\n",
      "Your job 813435 (\"ssp_cdivoire_raviart_114\") has been submitted\n",
      "Your job 813436 (\"ssp_cdivoire_raviart_89\") has been submitted\n",
      "Your job 813437 (\"ssp_cdivoire_raviart_90\") has been submitted\n",
      "Your job 813438 (\"ssp_cdivoire_raviart_91\") has been submitted\n",
      "Your job 813439 (\"ssp_cdivoire_raviart_92\") has been submitted\n",
      "Your job 813440 (\"ssp_cdivoire_raviart_99\") has been submitted\n",
      "Your job 813441 (\"ssp_niger_libore_156\") has been submitted\n",
      "Your job 813442 (\"ssp_niger_libore_157\") has been submitted\n",
      "Your job 813443 (\"ssp_niger_libore_159\") has been submitted\n",
      "Your job 813444 (\"ssp_niger_libore_165\") has been submitted\n",
      "Your job 813445 (\"ssp_niger_libore_166\") has been submitted\n",
      "Your job 813446 (\"ssp_niger_libore_167\") has been submitted\n",
      "Your job 813447 (\"ssp_niger_libore_168\") has been submitted\n",
      "Your job 813448 (\"ssp_niger_libore_169\") has been submitted\n",
      "Your job 813449 (\"ssp_niger_libore_173\") has been submitted\n",
      "Your job 813450 (\"ssp_zambia_kafue_71\") has been submitted\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "QSUB=\"qsub -V -cwd -S /bin/bash -q all.q -j y -p -1023 -pe smp 24\" \n",
    "REF=\"/master/nplatt/sch_hae_scan/data/GCF_000699445.3_UoM_Shae.V3_genomic.fna\"\n",
    "\n",
    "for SAMPLE in $(cat samples.list); do\n",
    "    \n",
    "    CRAM=\"/master/nplatt/sch_hae_scan/results/mapped_reads/${SAMPLE}_processed.cram\"\n",
    "    CMD=\"conda run -n popstructure --cwd . mosdepth --use-median --threads 4 --fasta ${REF} ${SAMPLE} ${CRAM}\"\n",
    "    \n",
    "    echo $CMD | $QSUB -N ${SAMPLE} -o ${SAMPLE}.log\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6c3b06a7-e725-40e7-8e78-3fb8f80ec2b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_covs=[]\n",
    "\n",
    "for sample in samples:\n",
    "    with open(f\"{sample}.mosdepth.summary.txt\", 'r') as f:\n",
    "        for line in f:\n",
    "            pass\n",
    "        last_line = line\n",
    "    \n",
    "        mean_covs.append(float(last_line.split(\"\\t\")[3]))      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "60ac5c3d-796a-4e64-a05d-30b548f3de73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29.7"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.DataFrame( {\n",
    "    \"sample\": samples,\n",
    "    \"cov\": mean_covs\n",
    "})\n",
    "\n",
    "df[\"cov\"].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "110619a2-ec07-41b4-8f4f-8bd94c86e3e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cov</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>20.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>23.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>23.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>24.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>27.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>31.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>39.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>50.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>41.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>43.49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>161 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample    cov\n",
       "0            Sb_NG_au_1.2  20.53\n",
       "1           Sb_NG_au_2.10  23.15\n",
       "2           Sb_NG_au_2.13  23.30\n",
       "3            Sb_NG_au_2.5  24.73\n",
       "4            Sb_NG_au_2.6  27.05\n",
       "..                    ...    ...\n",
       "166  ssp_niger_libore_167  31.23\n",
       "167  ssp_niger_libore_168  39.19\n",
       "168  ssp_niger_libore_169  50.42\n",
       "169  ssp_niger_libore_173  41.31\n",
       "170   ssp_zambia_kafue_71  43.49\n",
       "\n",
       "[161 rows x 2 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.to_csv(\"genome_cov.csv\", header=True, index=False, sep=\",\")\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "popstructure",
   "language": "python",
   "name": "popstructure"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
