{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ae6e2cb-8406-4873-96df-893bee13ace3",
   "metadata": {},
   "source": [
    "run with `popstrucure` env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "101a4f4b-10b0-4add-bad8-2effdecd0c72",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Unable to determine R home: [Errno 2] No such file or directory: 'R'\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import vcf\n",
    "from pathlib import Path\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import cm\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "import glob\n",
    "import numpy as np\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as mlines\n",
    "import statistics\n",
    "import allel\n",
    "from collections import defaultdict\n",
    "from itertools import combinations\n",
    "from haversine import haversine\n",
    "from skbio.stats.distance import mantel\n",
    "from scipy.stats import ttest_ind\n",
    "import cartopy.crs as ccrs\n",
    "import cartopy.feature as cf\n",
    "from pykrige.ok import OrdinaryKriging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fa2e0624-d18c-44e8-af03-8217a93dbcd2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_scan\"\n",
    "results_dir=\"{}/results\".format(proj_dir)\n",
    "\n",
    "os.chdir(proj_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4fc670a7-a613-4d36-b9d8-a130caf75cdc",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_df=pd.read_csv(\"{}/data/seq_and_sample_docs/all_sh_sb_sample_data.csv\".format(proj_dir), sep=\",\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c25fe0-dd98-4811-9088-754f12f6744e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Mapping rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2f4c6e69-e8eb-4c77-ac86-aab2dd9afe09",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"{}/mapping_rates\".format(results_dir)).mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(\"{}/mapping_rates\".format(results_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f55399e4-09fe-4a47-98bb-d65f81f19073",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_df=pd.read_csv(\"../initial_pca/pca_df.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1578027-11af-48a0-95e3-91399428b1fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>locale</th>\n",
       "      <th>host</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>project</th>\n",
       "      <th>data_type</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>kmeans_group</th>\n",
       "      <th>pca_species</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>-0.174693</td>\n",
       "      <td>0.025622</td>\n",
       "      <td>0.001506</td>\n",
       "      <td>0.004149</td>\n",
       "      <td>0.035574</td>\n",
       "      <td>0.055396</td>\n",
       "      <td>0.032718</td>\n",
       "      <td>-0.019520</td>\n",
       "      <td>0.007569</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>-0.173848</td>\n",
       "      <td>0.025251</td>\n",
       "      <td>0.001886</td>\n",
       "      <td>0.005161</td>\n",
       "      <td>0.036499</td>\n",
       "      <td>0.060013</td>\n",
       "      <td>0.032135</td>\n",
       "      <td>-0.019021</td>\n",
       "      <td>0.003526</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>-0.174493</td>\n",
       "      <td>0.025294</td>\n",
       "      <td>0.001699</td>\n",
       "      <td>0.001562</td>\n",
       "      <td>0.037332</td>\n",
       "      <td>0.060250</td>\n",
       "      <td>0.030712</td>\n",
       "      <td>-0.022199</td>\n",
       "      <td>0.003399</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>-0.173862</td>\n",
       "      <td>0.025171</td>\n",
       "      <td>0.003061</td>\n",
       "      <td>-0.001638</td>\n",
       "      <td>0.037943</td>\n",
       "      <td>0.059797</td>\n",
       "      <td>0.031391</td>\n",
       "      <td>-0.015780</td>\n",
       "      <td>0.000577</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>-0.173011</td>\n",
       "      <td>0.025618</td>\n",
       "      <td>0.001527</td>\n",
       "      <td>0.005455</td>\n",
       "      <td>0.038614</td>\n",
       "      <td>0.061741</td>\n",
       "      <td>0.032516</td>\n",
       "      <td>-0.022343</td>\n",
       "      <td>0.009809</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.023482</td>\n",
       "      <td>-0.078226</td>\n",
       "      <td>-0.011342</td>\n",
       "      <td>-0.059942</td>\n",
       "      <td>0.017520</td>\n",
       "      <td>-0.043266</td>\n",
       "      <td>0.041449</td>\n",
       "      <td>-0.016997</td>\n",
       "      <td>0.021684</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.022237</td>\n",
       "      <td>-0.077112</td>\n",
       "      <td>-0.019412</td>\n",
       "      <td>-0.059978</td>\n",
       "      <td>0.006374</td>\n",
       "      <td>-0.004766</td>\n",
       "      <td>0.030555</td>\n",
       "      <td>-0.041763</td>\n",
       "      <td>0.080463</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.023857</td>\n",
       "      <td>-0.073894</td>\n",
       "      <td>-0.020205</td>\n",
       "      <td>-0.050080</td>\n",
       "      <td>0.015119</td>\n",
       "      <td>-0.015182</td>\n",
       "      <td>0.038133</td>\n",
       "      <td>-0.019263</td>\n",
       "      <td>0.036911</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>0.026630</td>\n",
       "      <td>-0.083945</td>\n",
       "      <td>-0.003698</td>\n",
       "      <td>-0.069531</td>\n",
       "      <td>-0.009745</td>\n",
       "      <td>-0.085123</td>\n",
       "      <td>0.095988</td>\n",
       "      <td>-0.108937</td>\n",
       "      <td>0.082528</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>0.053261</td>\n",
       "      <td>0.112131</td>\n",
       "      <td>-0.096936</td>\n",
       "      <td>-0.011639</td>\n",
       "      <td>-0.039766</td>\n",
       "      <td>0.015451</td>\n",
       "      <td>0.005055</td>\n",
       "      <td>-0.002872</td>\n",
       "      <td>-0.008524</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-15.783442</td>\n",
       "      <td>28.172582</td>\n",
       "      <td>1</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_se</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample_name         1         2         3         4         5  \\\n",
       "0            Sb_NG_au_1.2 -0.174693  0.025622  0.001506  0.004149  0.035574   \n",
       "1           Sb_NG_au_2.10 -0.173848  0.025251  0.001886  0.005161  0.036499   \n",
       "2           Sb_NG_au_2.13 -0.174493  0.025294  0.001699  0.001562  0.037332   \n",
       "3            Sb_NG_au_2.5 -0.173862  0.025171  0.003061 -0.001638  0.037943   \n",
       "4            Sb_NG_au_2.6 -0.173011  0.025618  0.001527  0.005455  0.038614   \n",
       "..                    ...       ...       ...       ...       ...       ...   \n",
       "166  ssp_niger_libore_167  0.023482 -0.078226 -0.011342 -0.059942  0.017520   \n",
       "167  ssp_niger_libore_168  0.022237 -0.077112 -0.019412 -0.059978  0.006374   \n",
       "168  ssp_niger_libore_169  0.023857 -0.073894 -0.020205 -0.050080  0.015119   \n",
       "169  ssp_niger_libore_173  0.026630 -0.083945 -0.003698 -0.069531 -0.009745   \n",
       "170   ssp_zambia_kafue_71  0.053261  0.112131 -0.096936 -0.011639 -0.039766   \n",
       "\n",
       "            6         7         8         9  ...  locale   host  phenotype  \\\n",
       "0    0.055396  0.032718 -0.019520  0.007569  ...      na    cow      bovis   \n",
       "1    0.060013  0.032135 -0.019021  0.003526  ...      na    cow      bovis   \n",
       "2    0.060250  0.030712 -0.022199  0.003399  ...      na    cow      bovis   \n",
       "3    0.059797  0.031391 -0.015780  0.000577  ...      na    cow      bovis   \n",
       "4    0.061741  0.032516 -0.022343  0.009809  ...      na    cow      bovis   \n",
       "..        ...       ...       ...       ...  ...     ...    ...        ...   \n",
       "166 -0.043266  0.041449 -0.016997  0.021684  ...      na  snail        unk   \n",
       "167 -0.004766  0.030555 -0.041763  0.080463  ...      na  snail        unk   \n",
       "168 -0.015182  0.038133 -0.019263  0.036911  ...      na  snail        unk   \n",
       "169 -0.085123  0.095988 -0.108937  0.082528  ...      na  snail        unk   \n",
       "170  0.015451  0.005055 -0.002872 -0.008524  ...      na  snail        unk   \n",
       "\n",
       "         project  data_type        lat        lon  kmeans_group  \\\n",
       "0        nigeria     genome   7.067448   6.271139             2   \n",
       "1        nigeria     genome   7.067448   6.271139             2   \n",
       "2        nigeria     genome   7.067448   6.271139             2   \n",
       "3        nigeria     genome   7.067448   6.271139             2   \n",
       "4        nigeria     genome   7.067448   6.271139             2   \n",
       "..           ...        ...        ...        ...           ...   \n",
       "166  scan_africa     genome  13.390080   2.197110             0   \n",
       "167  scan_africa     genome  13.390080   2.197110             0   \n",
       "168  scan_africa     genome  13.404170   2.188020             0   \n",
       "169  scan_africa     genome  13.404170   2.188020             0   \n",
       "170  scan_africa     genome -15.783442  28.172582             1   \n",
       "\n",
       "       pca_species  kmeans_label  \n",
       "0          s_bovis            sb  \n",
       "1          s_bovis            sb  \n",
       "2          s_bovis            sb  \n",
       "3          s_bovis            sb  \n",
       "4          s_bovis            sb  \n",
       "..             ...           ...  \n",
       "166  s_haematobium         sh_nw  \n",
       "167  s_haematobium         sh_nw  \n",
       "168  s_haematobium         sh_nw  \n",
       "169  s_haematobium         sh_nw  \n",
       "170  s_haematobium         sh_se  \n",
       "\n",
       "[171 rows x 37 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "40bff22f-f145-4bff-897b-f2d015e28cfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>map_rate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bovis_ethiopia_SRR13579874</td>\n",
       "      <td>0.7875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bovis_keyna_SRR13579878</td>\n",
       "      <td>0.7666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bovis_nigeria_SRR13579877</td>\n",
       "      <td>0.7650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bovis_senegal_SRR13579876</td>\n",
       "      <td>0.7638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bovis_spain_SRR13579873</td>\n",
       "      <td>0.7635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.8637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.8970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.8977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>0.8908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>0.8743</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>211 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample_name  map_rate\n",
       "0    bovis_ethiopia_SRR13579874    0.7875\n",
       "1       bovis_keyna_SRR13579878    0.7666\n",
       "2     bovis_nigeria_SRR13579877    0.7650\n",
       "3     bovis_senegal_SRR13579876    0.7638\n",
       "4       bovis_spain_SRR13579873    0.7635\n",
       "..                          ...       ...\n",
       "206        ssp_niger_libore_167    0.8637\n",
       "207        ssp_niger_libore_168    0.8970\n",
       "208        ssp_niger_libore_169    0.8977\n",
       "209        ssp_niger_libore_173    0.8908\n",
       "210         ssp_zambia_kafue_71    0.8743\n",
       "\n",
       "[211 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_df = pd.read_csv(f\"{results_dir}/mapped_reads/mapping_rates.csv\", sep=\",\", header=None)\n",
    "map_df.columns=[\"sample_name\", \"map_rate\"]\n",
    "map_df[\"map_rate\"] = map_df['map_rate'].str.rstrip(\"%\").astype(float)/100\n",
    "map_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ea04a219-8efc-4380-b5b5-19e3a253aa8c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>map_rate</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bovis_ethiopia_SRR13579874</td>\n",
       "      <td>0.7875</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bovis_keyna_SRR13579878</td>\n",
       "      <td>0.7666</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bovis_senegal_SRR13579876</td>\n",
       "      <td>0.7638</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bovis_tanzania_ERR103048</td>\n",
       "      <td>0.7918</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bovis_tanzania_SRR7867225</td>\n",
       "      <td>0.7755</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.8637</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.8970</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.8977</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>0.8908</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>0.8743</td>\n",
       "      <td>sh_se</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample_name  map_rate kmeans_label\n",
       "0    bovis_ethiopia_SRR13579874    0.7875           sb\n",
       "1       bovis_keyna_SRR13579878    0.7666           sb\n",
       "2     bovis_senegal_SRR13579876    0.7638           sb\n",
       "3      bovis_tanzania_ERR103048    0.7918           sb\n",
       "4     bovis_tanzania_SRR7867225    0.7755           sb\n",
       "..                          ...       ...          ...\n",
       "166        ssp_niger_libore_167    0.8637        sh_nw\n",
       "167        ssp_niger_libore_168    0.8970        sh_nw\n",
       "168        ssp_niger_libore_169    0.8977        sh_nw\n",
       "169        ssp_niger_libore_173    0.8908        sh_nw\n",
       "170         ssp_zambia_kafue_71    0.8743        sh_se\n",
       "\n",
       "[171 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_df = pd.merge(map_df, pca_df[[\"sample_name\", \"kmeans_label\"]], on='sample_name', how='outer').dropna().reset_index(drop=True)\n",
    "map_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cecda256-f35e-459a-b10c-dbf1713c1a1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_288398/3888416185.py:1: FutureWarning: ['sample_name'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.\n",
      "  map_df.groupby(by=[\"kmeans_label\"]).agg(['mean', 'count', 'std'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"3\" halign=\"left\">map_rate</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>count</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kmeans_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s_guineensis</th>\n",
       "      <td>0.607500</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_intercalatum</th>\n",
       "      <td>0.727600</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_margrebowiei</th>\n",
       "      <td>0.674500</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_matthei</th>\n",
       "      <td>0.704400</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sb</th>\n",
       "      <td>0.776565</td>\n",
       "      <td>23</td>\n",
       "      <td>0.080901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sh_nw</th>\n",
       "      <td>0.826742</td>\n",
       "      <td>84</td>\n",
       "      <td>0.093948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sh_se</th>\n",
       "      <td>0.763060</td>\n",
       "      <td>60</td>\n",
       "      <td>0.142906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                map_rate                \n",
       "                    mean count       std\n",
       "kmeans_label                            \n",
       "s_guineensis    0.607500     1       NaN\n",
       "s_intercalatum  0.727600     1       NaN\n",
       "s_margrebowiei  0.674500     1       NaN\n",
       "s_matthei       0.704400     1       NaN\n",
       "sb              0.776565    23  0.080901\n",
       "sh_nw           0.826742    84  0.093948\n",
       "sh_se           0.763060    60  0.142906"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_df.groupby(by=[\"kmeans_label\"]).agg(['mean', 'count', 'std'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c82b7174-a73d-44f8-bd43-ce600797a1bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=0.8499875241109823, pvalue=0.3966021310889227)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#define samples\n",
    "sh = map_df[map_df['kmeans_label'].isin([\"sh_nw\", \"sh_se\"])]\n",
    "sb = map_df[map_df['kmeans_label']=='sb']\n",
    "\n",
    "#perform independent two sample t-test\n",
    "ttest_ind(sh['map_rate'], sb['map_rate'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "ea1ceffe-9d64-4b12-8e2e-72b175a4ec95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3256551</td>\n",
       "      <td>0.085253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>4007156</td>\n",
       "      <td>0.104904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3597035</td>\n",
       "      <td>0.094167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3738947</td>\n",
       "      <td>0.097882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3616554</td>\n",
       "      <td>0.094678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>8932549</td>\n",
       "      <td>0.233846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3975609</td>\n",
       "      <td>0.104078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>597245</td>\n",
       "      <td>0.015635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>1491990</td>\n",
       "      <td>0.039059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>2485544</td>\n",
       "      <td>0.065069</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     INDV    N_DATA  N_GENOTYPES_FILTERED   N_MISS    F_MISS\n",
       "0            Sb_NG_ak_2.1  38198427                     0  3256551  0.085253\n",
       "1            Sb_NG_au_1.2  38198427                     0  4007156  0.104904\n",
       "2           Sb_NG_au_2.10  38198427                     0  3597035  0.094167\n",
       "3           Sb_NG_au_2.13  38198427                     0  3738947  0.097882\n",
       "4            Sb_NG_au_2.5  38198427                     0  3616554  0.094678\n",
       "..                    ...       ...                   ...      ...       ...\n",
       "214  ssp_niger_libore_167  38198427                     0  8932549  0.233846\n",
       "215  ssp_niger_libore_168  38198427                     0  3975609  0.104078\n",
       "216  ssp_niger_libore_169  38198427                     0   597245  0.015635\n",
       "217  ssp_niger_libore_173  38198427                     0  1491990  0.039059\n",
       "218   ssp_zambia_kafue_71  38198427                     0  2485544  0.065069\n",
       "\n",
       "[219 rows x 5 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idv_df=pd.read_csv(\"../filter_genotypes/indv_gt_freq.tbl\", sep=\"\\t\")\n",
    "idv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "f50a8a24-202e-4b11-9463-7b6f6ebd3ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>4007156</td>\n",
       "      <td>0.104904</td>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3597035</td>\n",
       "      <td>0.094167</td>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3738947</td>\n",
       "      <td>0.097882</td>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3616554</td>\n",
       "      <td>0.094678</td>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3299502</td>\n",
       "      <td>0.086378</td>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>8932549</td>\n",
       "      <td>0.233846</td>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>3975609</td>\n",
       "      <td>0.104078</td>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>597245</td>\n",
       "      <td>0.015635</td>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>1491990</td>\n",
       "      <td>0.039059</td>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>38198427</td>\n",
       "      <td>0</td>\n",
       "      <td>2485544</td>\n",
       "      <td>0.065069</td>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>sh_se</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     INDV    N_DATA  N_GENOTYPES_FILTERED   N_MISS    F_MISS  \\\n",
       "0            Sb_NG_au_1.2  38198427                     0  4007156  0.104904   \n",
       "1           Sb_NG_au_2.10  38198427                     0  3597035  0.094167   \n",
       "2           Sb_NG_au_2.13  38198427                     0  3738947  0.097882   \n",
       "3            Sb_NG_au_2.5  38198427                     0  3616554  0.094678   \n",
       "4            Sb_NG_au_2.6  38198427                     0  3299502  0.086378   \n",
       "..                    ...       ...                   ...      ...       ...   \n",
       "166  ssp_niger_libore_167  38198427                     0  8932549  0.233846   \n",
       "167  ssp_niger_libore_168  38198427                     0  3975609  0.104078   \n",
       "168  ssp_niger_libore_169  38198427                     0   597245  0.015635   \n",
       "169  ssp_niger_libore_173  38198427                     0  1491990  0.039059   \n",
       "170   ssp_zambia_kafue_71  38198427                     0  2485544  0.065069   \n",
       "\n",
       "              sample_name kmeans_label  \n",
       "0            Sb_NG_au_1.2           sb  \n",
       "1           Sb_NG_au_2.10           sb  \n",
       "2           Sb_NG_au_2.13           sb  \n",
       "3            Sb_NG_au_2.5           sb  \n",
       "4            Sb_NG_au_2.6           sb  \n",
       "..                    ...          ...  \n",
       "166  ssp_niger_libore_167        sh_nw  \n",
       "167  ssp_niger_libore_168        sh_nw  \n",
       "168  ssp_niger_libore_169        sh_nw  \n",
       "169  ssp_niger_libore_173        sh_nw  \n",
       "170   ssp_zambia_kafue_71        sh_se  \n",
       "\n",
       "[171 rows x 7 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idv_df = pd.merge(idv_df, pca_df[[\"sample_name\", \"kmeans_label\"]], left_on='INDV', right_on='sample_name', how='outer').dropna().reset_index(drop=True)\n",
    "idv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7aa75978-c738-4523-904b-afb0224282e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean</th>\n",
       "      <th>count</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kmeans_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>s_guineensis</th>\n",
       "      <td>0.376764</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_intercalatum</th>\n",
       "      <td>0.083733</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_margrebowiei</th>\n",
       "      <td>0.231996</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_matthei</th>\n",
       "      <td>0.100289</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sb</th>\n",
       "      <td>0.161157</td>\n",
       "      <td>23</td>\n",
       "      <td>0.105172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sh_nw</th>\n",
       "      <td>0.154415</td>\n",
       "      <td>84</td>\n",
       "      <td>0.129582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sh_se</th>\n",
       "      <td>0.140557</td>\n",
       "      <td>60</td>\n",
       "      <td>0.125487</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    mean  count       std\n",
       "kmeans_label                             \n",
       "s_guineensis    0.376764      1       NaN\n",
       "s_intercalatum  0.083733      1       NaN\n",
       "s_margrebowiei  0.231996      1       NaN\n",
       "s_matthei       0.100289      1       NaN\n",
       "sb              0.161157     23  0.105172\n",
       "sh_nw           0.154415     84  0.129582\n",
       "sh_se           0.140557     60  0.125487"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idv_df.groupby(by=[\"kmeans_label\"])[\"F_MISS\"].agg(['mean', 'count', 'std'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "89d5bf9a-1409-4ce7-86a3-fa2b40d8d340",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Ttest_indResult(statistic=-0.4463605776958125, pvalue=0.6559220423447343)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#define samples\n",
    "sh = idv_df[idv_df['kmeans_label'].isin([\"sh_nw\", \"sh_se\"])]\n",
    "sb = idv_df[idv_df['kmeans_label']=='sb']\n",
    "\n",
    "#perform independent two sample t-test\n",
    "ttest_ind(sh['F_MISS'], sb['F_MISS'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "popstructure",
   "language": "python",
   "name": "popstructure"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
