{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b431a647-9588-40ac-9abd-d063d1120145",
   "metadata": {},
   "outputs": [],
   "source": [
    "#conda activate scan_snpeff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5c3b33dc-894a-45ab-b98e-c06dc666896a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ecc5bc8-01e3-4e2a-906f-a6ce99b645ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_scan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d82e2ec7-6f31-4296-b04c-20ddfb6112a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"{}/results/snpeff\".format(proj_dir)).mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(\"{}/results/snpeff\".format(proj_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65a5ad39-e014-4ff0-8bf5-d8daa6eaedd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "conda run -n scan_snpeff --cwd . --live-stream \\\n",
    "    java -jar snpEff.jar build -noCheckCds -noCheckProtein -gff3 -v GCF_000699445.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9db8cf6-2053-48cd-8ab1-77ac9ab161e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "conda run -n scan_snpeff --cwd . --live-stream \\\n",
    "    java -jar -Xmx64g snpEff.jar \\\n",
    "        ann \\\n",
    "        -csvStats snpeff_stats.csv \\\n",
    "        -i vcf \\\n",
    "        -o vcf \\\n",
    "        -no-downstream \\\n",
    "        -no-intergenic \\\n",
    "        -no-intron \\\n",
    "        -no-upstream \\\n",
    "        -fastaProt snpeff_prots.fas \\\n",
    "        GCF_000699445.3 \\\n",
    "        ../phasing/beagle.vcf \\\n",
    "        >snpeff.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7561e3f5-3396-4a89-90f5-323b86b2d254",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>locale</th>\n",
       "      <th>host</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>project</th>\n",
       "      <th>data_type</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>kmeans_group</th>\n",
       "      <th>pca_species</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_au_1.2</td>\n",
       "      <td>-0.186571</td>\n",
       "      <td>0.028344</td>\n",
       "      <td>0.002659</td>\n",
       "      <td>0.004098</td>\n",
       "      <td>-0.070428</td>\n",
       "      <td>-0.050869</td>\n",
       "      <td>0.009811</td>\n",
       "      <td>-0.013460</td>\n",
       "      <td>-0.016134</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_au_2.10</td>\n",
       "      <td>-0.186390</td>\n",
       "      <td>0.027938</td>\n",
       "      <td>0.002539</td>\n",
       "      <td>0.005840</td>\n",
       "      <td>-0.070521</td>\n",
       "      <td>-0.050565</td>\n",
       "      <td>0.007871</td>\n",
       "      <td>-0.013283</td>\n",
       "      <td>-0.014293</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_au_2.13</td>\n",
       "      <td>-0.186938</td>\n",
       "      <td>0.028149</td>\n",
       "      <td>0.002587</td>\n",
       "      <td>0.002080</td>\n",
       "      <td>-0.073837</td>\n",
       "      <td>-0.052819</td>\n",
       "      <td>0.006287</td>\n",
       "      <td>-0.017596</td>\n",
       "      <td>-0.015425</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_au_2.5</td>\n",
       "      <td>-0.186747</td>\n",
       "      <td>0.027939</td>\n",
       "      <td>0.003308</td>\n",
       "      <td>-0.001183</td>\n",
       "      <td>-0.072298</td>\n",
       "      <td>-0.050190</td>\n",
       "      <td>0.007316</td>\n",
       "      <td>-0.011898</td>\n",
       "      <td>-0.008823</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_au_2.6</td>\n",
       "      <td>-0.186556</td>\n",
       "      <td>0.028644</td>\n",
       "      <td>0.001854</td>\n",
       "      <td>0.006496</td>\n",
       "      <td>-0.069384</td>\n",
       "      <td>-0.049374</td>\n",
       "      <td>0.007875</td>\n",
       "      <td>-0.015933</td>\n",
       "      <td>-0.022326</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>cow</td>\n",
       "      <td>bovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.067448</td>\n",
       "      <td>6.271139</td>\n",
       "      <td>2</td>\n",
       "      <td>s_bovis</td>\n",
       "      <td>sb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>ssp_niger_libore_166</td>\n",
       "      <td>0.021444</td>\n",
       "      <td>-0.079803</td>\n",
       "      <td>-0.018635</td>\n",
       "      <td>-0.066947</td>\n",
       "      <td>0.003088</td>\n",
       "      <td>0.018973</td>\n",
       "      <td>0.086455</td>\n",
       "      <td>-0.034980</td>\n",
       "      <td>-0.046315</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.388050</td>\n",
       "      <td>2.199130</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.022669</td>\n",
       "      <td>-0.077370</td>\n",
       "      <td>-0.009122</td>\n",
       "      <td>-0.064440</td>\n",
       "      <td>-0.003224</td>\n",
       "      <td>0.026673</td>\n",
       "      <td>0.059154</td>\n",
       "      <td>0.003149</td>\n",
       "      <td>-0.002788</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.020863</td>\n",
       "      <td>-0.077872</td>\n",
       "      <td>-0.016320</td>\n",
       "      <td>-0.064476</td>\n",
       "      <td>-0.005623</td>\n",
       "      <td>-0.001633</td>\n",
       "      <td>0.037675</td>\n",
       "      <td>-0.009439</td>\n",
       "      <td>-0.063435</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.022264</td>\n",
       "      <td>-0.076285</td>\n",
       "      <td>-0.019965</td>\n",
       "      <td>-0.053381</td>\n",
       "      <td>-0.012734</td>\n",
       "      <td>0.007188</td>\n",
       "      <td>0.066627</td>\n",
       "      <td>0.009003</td>\n",
       "      <td>-0.007182</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>ssp_zambia_kafue_71</td>\n",
       "      <td>0.051109</td>\n",
       "      <td>0.115655</td>\n",
       "      <td>-0.100785</td>\n",
       "      <td>-0.014122</td>\n",
       "      <td>0.026974</td>\n",
       "      <td>-0.037202</td>\n",
       "      <td>0.001224</td>\n",
       "      <td>-0.005005</td>\n",
       "      <td>0.008020</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-15.783442</td>\n",
       "      <td>28.172582</td>\n",
       "      <td>1</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_se</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>166 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample_name         1         2         3         4         5  \\\n",
       "0            Sb_NG_au_1.2 -0.186571  0.028344  0.002659  0.004098 -0.070428   \n",
       "1           Sb_NG_au_2.10 -0.186390  0.027938  0.002539  0.005840 -0.070521   \n",
       "2           Sb_NG_au_2.13 -0.186938  0.028149  0.002587  0.002080 -0.073837   \n",
       "3            Sb_NG_au_2.5 -0.186747  0.027939  0.003308 -0.001183 -0.072298   \n",
       "4            Sb_NG_au_2.6 -0.186556  0.028644  0.001854  0.006496 -0.069384   \n",
       "..                    ...       ...       ...       ...       ...       ...   \n",
       "161  ssp_niger_libore_166  0.021444 -0.079803 -0.018635 -0.066947  0.003088   \n",
       "162  ssp_niger_libore_167  0.022669 -0.077370 -0.009122 -0.064440 -0.003224   \n",
       "163  ssp_niger_libore_168  0.020863 -0.077872 -0.016320 -0.064476 -0.005623   \n",
       "164  ssp_niger_libore_169  0.022264 -0.076285 -0.019965 -0.053381 -0.012734   \n",
       "165   ssp_zambia_kafue_71  0.051109  0.115655 -0.100785 -0.014122  0.026974   \n",
       "\n",
       "            6         7         8         9  ...  locale   host  phenotype  \\\n",
       "0   -0.050869  0.009811 -0.013460 -0.016134  ...      na    cow      bovis   \n",
       "1   -0.050565  0.007871 -0.013283 -0.014293  ...      na    cow      bovis   \n",
       "2   -0.052819  0.006287 -0.017596 -0.015425  ...      na    cow      bovis   \n",
       "3   -0.050190  0.007316 -0.011898 -0.008823  ...      na    cow      bovis   \n",
       "4   -0.049374  0.007875 -0.015933 -0.022326  ...      na    cow      bovis   \n",
       "..        ...       ...       ...       ...  ...     ...    ...        ...   \n",
       "161  0.018973  0.086455 -0.034980 -0.046315  ...      na  snail        unk   \n",
       "162  0.026673  0.059154  0.003149 -0.002788  ...      na  snail        unk   \n",
       "163 -0.001633  0.037675 -0.009439 -0.063435  ...      na  snail        unk   \n",
       "164  0.007188  0.066627  0.009003 -0.007182  ...      na  snail        unk   \n",
       "165 -0.037202  0.001224 -0.005005  0.008020  ...      na  snail        unk   \n",
       "\n",
       "         project  data_type        lat        lon  kmeans_group  \\\n",
       "0        nigeria     genome   7.067448   6.271139             2   \n",
       "1        nigeria     genome   7.067448   6.271139             2   \n",
       "2        nigeria     genome   7.067448   6.271139             2   \n",
       "3        nigeria     genome   7.067448   6.271139             2   \n",
       "4        nigeria     genome   7.067448   6.271139             2   \n",
       "..           ...        ...        ...        ...           ...   \n",
       "161  scan_africa     genome  13.388050   2.199130             0   \n",
       "162  scan_africa     genome  13.390080   2.197110             0   \n",
       "163  scan_africa     genome  13.390080   2.197110             0   \n",
       "164  scan_africa     genome  13.404170   2.188020             0   \n",
       "165  scan_africa     genome -15.783442  28.172582             1   \n",
       "\n",
       "       pca_species  kmeans_label  \n",
       "0          s_bovis            sb  \n",
       "1          s_bovis            sb  \n",
       "2          s_bovis            sb  \n",
       "3          s_bovis            sb  \n",
       "4          s_bovis            sb  \n",
       "..             ...           ...  \n",
       "161  s_haematobium         sh_nw  \n",
       "162  s_haematobium         sh_nw  \n",
       "163  s_haematobium         sh_nw  \n",
       "164  s_haematobium         sh_nw  \n",
       "165  s_haematobium         sh_se  \n",
       "\n",
       "[166 rows x 37 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_df=pd.read_csv(f\"{proj_dir}/results/pca/pca_df.csv\", sep=\",\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cda17926-341c-4184-ac59-1edb7d88d087",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the genetic data\n",
    "vcf_reader = vcf.Reader(open('snpeff.vcf', 'r'))\n",
    "#get the sample order\n",
    "samples=vcf_reader.samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a5ffe908-5489-43de-b354-ad2cae39f128",
   "metadata": {},
   "outputs": [],
   "source": [
    "pops=[\"sb\", \"sh_nw\", \"sh_se\"]\n",
    "\n",
    "for pop in pops:\n",
    "    a = pca_df.loc[ pca_df[\"kmeans_label\"] == pop, \"sample_name\"].values\n",
    "    np.savetxt(f\"{pop}.list\", a, fmt=\"%s\")\n",
    "    \n",
    "    cmd = f\"vcftools --vcf snpeff_beagle.vcf --bed ../rfmix/merged_high_sb_regions.bed --freq --keep {pop}.list --stdout >{pop}.high_sb.freq\"\n",
    "    \n",
    "    !conda run -n scan_snpeff --cwd . --live-stream {cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a24efb-8fc3-46ed-a8b2-545b9bebeb9c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8eb5756-beb6-40fb-a97e-4175c30ba904",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da8cb997-df7e-4e93-bd11-b9d9b332f8b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03fd150f-12f9-4839-9bb4-3a27bea58b77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65bf345d-c525-4be1-bd7d-6b497cb7c23a",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf snpeff_beagle.vcf \\\n",
    "    --bed merged_high_sb_regions.bed \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >peak.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3b5063c-5574-48bb-9d64-f59bbe73405f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf snpeff_beagle.vcf \\\n",
    "    --bed merged_high_sb_regions.bed \\\n",
    "    --freq \\\n",
    "    --keep <SB> \\\n",
    "    --stdout \\\n",
    "    >sb_high_sb.freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe7c1277-ca1f-4743-ba48-0c1dfcd4ef26",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf snpeff_beagle.vcf \\\n",
    "    --bed merged_high_sb_regions.bed \\\n",
    "    --freq \\\n",
    "    --keep <SB> \\\n",
    "    --stdout \\\n",
    "    >peak.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ca3512-b60a-4c3f-bbf3-c4e6f784f5aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf snpeff_beagle.vcf \\\n",
    "    --bed merged_high_sb_regions.bed \\\n",
    "    --freq \\\n",
    "    --keep <SB> \\\n",
    "    --stdout \\\n",
    "    >peak.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34088567-d2c3-4d6d-88bc-9776604edbcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "MS3_00007802"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d227004-b8f4-44ac-b068-62e7c3024095",
   "metadata": {},
   "source": [
    "`hypothetical protein MS3_00007802 [Schistosoma haematobium]` on NCBI has the same AA seq as `MS3_0012553 Leishmanolysin-like peptidase` on parasite. (See below). But these appear to occur at different positions on the genome. Likely due to a difference in assemblies?\n",
    "\n",
    "- MS3_00007802 = scaffold020018:8,567,422-8,600,894 (which is chr4)\n",
    "- MS3_0012553 = CM039146.1:28478765-28489490 (which is chr4)\n",
    "\n",
    "```>aa\n",
    "MWKIKDIILFPIFGVHFIISGLRYDRFSTLCHHGVINHSEVVTHVVSSDVQKLIRRNSQT\n",
    "ELRIHTHYDRSIKSLKNFQDIKRDVIEVAIEFWEDALLLRQSHTVPRRLTLDRVCVDGAT\n",
    "KLMTFDNQLLTFCVNGCVEWTKCGPINIPVSHLNCRFIYNGTSMISGRKGEGIYRANFIL\n",
    "YISSLRTHRCNTAKVLGYAAHCQLEANTDRPMAGYINFCPDTLSNEYNDKMRSLFVATHE\n",
    "ILHSLGFSTSLFAFYRDKQNRPLTPRDPITFKPALGWYSGKNGQVYQWSDNVVRSVNRTW\n",
    "LSAFGTFQKLAHIVVLPTVVRIARSFYNCPNLDGVELEDEDDAGVYLTHWEKRLLENDLM\n",
    "TATYTNSFRISPITLAMMEDTGGHPIAPFCQQLTSSFHNKNNGLQVSCTPGEESYGFCNL\n",
    "IEYSKPLPDEYVYFVNTNFSTVTLNSEINLGTGNLTNQYPLVKLGGKIALANYCPYHQEI\n",
    "EWEDDVGKSIANTHCHASTNLKDPHHNFKLERFGMDSVCVEHSPNWYLINGDSLFVPPVE\n",
    "GAGCYTFRCSMTEGGLILELAGGMSIPCEVPGELSEINACLTKQSLTVKGKLVCPDCRLL\n",
    "CPLSECPKIYSISQNNVTYLRSNYSNDRSSSFLYETNMFIPKSRTIQYSQLIVNETESIV\n",
    "ENPLYIVENSKNSKGIVIQQAILPGACLVNIAYSSTSLTFSIFIYFVVFVQLFCL```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61467b2d-3f45-4e4c-8cf9-efc757ea7ad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#how many of these genes in the high_sb_regions have non-synonymous changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbc16d0f-7b04-4c14-a58d-21e14bfb6e36",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf snpeff.vcf \\\n",
    "    --chr NC_067199.1 \\\n",
    "    --from-bp 28466881 \\\n",
    "    --to-bp 28497752 \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >invad.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7061a2aa-4d81-42cf-998e-650f3e566e7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcftools \\\n",
    "    --vcf invad.vcf \\\n",
    "    --keep nw.list \\\n",
    "    --freq \\\n",
    "    --stdout \\\n",
    "    >nw.freq"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scan_snpeff",
   "language": "python",
   "name": "scan_snpeff"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
