{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "18c525a1-edd4-4dc8-8eab-964ce7433548",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "from pathlib import Path\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d768c55-7277-48a5-aa86-5b14cd57ff6b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_scan\"\n",
    "results_dir=f\"{proj_dir}/results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "296ba063-8043-484a-8e80-f02328dabef2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Path(f\"{results_dir}/outlier_table\").mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(f\"{results_dir}/outlier_table\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2d8b1bd-7d5b-4105-b89e-4e0ee69129fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#split genome into 10kb windows\n",
    "!conda run -n popstructure bedtools makewindows -w 337000 -s 3370 -g ../../data/GCF_000699445.3_UoM_Shae.V3_genomic.fna.fai >1p_windows.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3f3c3151-a056-4ea2-be91-cd7e946a5377",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda run -n popstructure bedtools intersect -c -a 1p_windows.bed -b ../post_phase_filtering/chrs_unrelated.vcf >n_snvs_per_window.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "549eb771-03e6-4f6c-a17d-2e0eac3fc5f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_snvs_df = pd.read_csv(\"n_snvs_per_window.bed\", sep=\"\\t\", header=None)\n",
    "n_snvs_df.columns=[\"chrom\", \"start\", \"stop\", \"n_snvs\"]\n",
    "filt_snvs_df = n_snvs_df.loc[n_snvs_df[\"n_snvs\"] > 10]\n",
    "\n",
    "filt_snvs_bed_df = filt_snvs_df[[\"chrom\", \"start\", \"stop\"]]\n",
    "filt_snvs_bed_df.to_csv(\"n_snvs_gt10.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f806efc-b466-4184-b967-21fb8d1f95a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pi_df=pd.read_csv(f\"{results_dir}/summary_stats/summary_stats.csv\", header=0, sep=\",\")\n",
    "# pi_df = pi_df.loc[pi_df[\"n_snps\"]>10]\n",
    "# pi_df[\"se_pi\"] = pi_df[\"se_pi\"] + 1e-7\n",
    "# pi_df[\"nw_pi\"] = pi_df[\"nw_pi\"] + 1e-7\n",
    "# pi_df[\"pi_ratio_nw_se\"] = np.log10(pi_df[\"nw_pi\"]/pi_df[\"se_pi\"])\n",
    "\n",
    "# rank = pi_df[\"pi_ratio_nw_se\"].rank() / np.isfinite(pi_df[\"pi_ratio_nw_se\"]).sum()\n",
    "# pi_df[\"rank\"] = rank\n",
    "# sig_pi_df = pi_df.loc[(rank <= 0.05)]\n",
    "\n",
    "\n",
    "# pi_bed_df = sig_pi_df[[\"chrom\", \"start\", \"stop\"]]\n",
    "# pi_bed_df[\"start\"] = pi_bed_df[\"start\"].astype(int)\n",
    "# pi_bed_df[\"stop\"] = pi_bed_df[\"stop\"].astype(int)\n",
    "\n",
    "# pi_bed_df.to_csv(\"pi_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "04acd2a2-c7b5-4346-945d-5247fe4d975c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pbs_df = pd.read_csv(f\"{results_dir}/pbs/pbs_nonphased.csv\", sep=\",\", header=0)\n",
    "\n",
    "# sig_pbs_df = pbs_df[pbs_df['significant']]\n",
    "\n",
    "# pbs_bed_df = sig_pbs_df[[\"chrom\", \"start\", \"stop\"]]\n",
    "# pbs_bed_df.to_csv(\"pbs_nonphased_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "129d8491-132d-4294-a162-2b112b544d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_df=pd.read_csv(f\"{results_dir}/selscan/sel_df_norm.csv\", sep=\",\", header=0)\n",
    "\n",
    "\n",
    "alpha = 0.05\n",
    "sig_bonf = sel_df[sel_df[\"p_value_bonferroni\"] <= alpha]\n",
    "\n",
    "sel_bed_df = sig_bonf[[\"chr\", \"pos\", \"pos\"]]\n",
    "sel_bed_df.to_csv(\"normxpehh_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "0a94b893-25bc-4d0d-a5cf-18d5f1000cb2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>gpos</th>\n",
       "      <th>p1</th>\n",
       "      <th>ihh1</th>\n",
       "      <th>p2</th>\n",
       "      <th>ihh2</th>\n",
       "      <th>xpehh</th>\n",
       "      <th>normxpehh</th>\n",
       "      <th>crit</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>color</th>\n",
       "      <th>norm_p_value</th>\n",
       "      <th>p_value_bonferroni</th>\n",
       "      <th>fdr</th>\n",
       "      <th>nw_intro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>147067</td>\n",
       "      <td>NC_067196.1:147067</td>\n",
       "      <td>147067.0</td>\n",
       "      <td>0.158537</td>\n",
       "      <td>3849.020</td>\n",
       "      <td>0.381356</td>\n",
       "      <td>2542.010</td>\n",
       "      <td>0.180172</td>\n",
       "      <td>1.58557</td>\n",
       "      <td>0</td>\n",
       "      <td>147067</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.112837</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.823171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>147068</td>\n",
       "      <td>NC_067196.1:147068</td>\n",
       "      <td>147068.0</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>3270.680</td>\n",
       "      <td>0.186441</td>\n",
       "      <td>1366.650</td>\n",
       "      <td>0.378980</td>\n",
       "      <td>2.42020</td>\n",
       "      <td>1</td>\n",
       "      <td>147068</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.015512</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.562670</td>\n",
       "      <td>0.823171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>147069</td>\n",
       "      <td>NC_067196.1:147069</td>\n",
       "      <td>147069.0</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>3022.890</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1374.130</td>\n",
       "      <td>0.342395</td>\n",
       "      <td>2.26661</td>\n",
       "      <td>1</td>\n",
       "      <td>147069</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.023414</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.679088</td>\n",
       "      <td>0.823171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>147070</td>\n",
       "      <td>NC_067196.1:147070</td>\n",
       "      <td>147070.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3023.680</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1374.130</td>\n",
       "      <td>0.342510</td>\n",
       "      <td>2.26709</td>\n",
       "      <td>1</td>\n",
       "      <td>147070</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.023385</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.678738</td>\n",
       "      <td>0.823171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>147072</td>\n",
       "      <td>NC_067196.1:147072</td>\n",
       "      <td>147072.0</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>3023.510</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1374.130</td>\n",
       "      <td>0.342485</td>\n",
       "      <td>2.26699</td>\n",
       "      <td>1</td>\n",
       "      <td>147072</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.023391</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.678825</td>\n",
       "      <td>0.823171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28585158</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477723</td>\n",
       "      <td>NC_067202.1:19477723</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>398.509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.169044</td>\n",
       "      <td>1.53885</td>\n",
       "      <td>0</td>\n",
       "      <td>317910191</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.123841</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.018293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28585159</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477730</td>\n",
       "      <td>NC_067202.1:19477730</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.036585</td>\n",
       "      <td>396.811</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.167189</td>\n",
       "      <td>1.53106</td>\n",
       "      <td>0</td>\n",
       "      <td>317910198</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.125755</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.018293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28585160</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477740</td>\n",
       "      <td>NC_067202.1:19477740</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>409.008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.180337</td>\n",
       "      <td>1.58626</td>\n",
       "      <td>0</td>\n",
       "      <td>317910208</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.112680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.018293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28585161</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477770</td>\n",
       "      <td>NC_067202.1:19477770</td>\n",
       "      <td>19477800.0</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>333.042</td>\n",
       "      <td>0.245763</td>\n",
       "      <td>207.029</td>\n",
       "      <td>0.206467</td>\n",
       "      <td>1.69596</td>\n",
       "      <td>0</td>\n",
       "      <td>317910238</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.089893</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.018293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28585162</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477785</td>\n",
       "      <td>NC_067202.1:19477785</td>\n",
       "      <td>19477800.0</td>\n",
       "      <td>0.012195</td>\n",
       "      <td>379.232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>236.080</td>\n",
       "      <td>0.205845</td>\n",
       "      <td>1.69335</td>\n",
       "      <td>0</td>\n",
       "      <td>317910253</td>\n",
       "      <td>#078d70</td>\n",
       "      <td>0.090389</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.018293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28585163 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  chr       pos                    id        gpos        p1  \\\n",
       "0         NC_067196.1    147067    NC_067196.1:147067    147067.0  0.158537   \n",
       "1         NC_067196.1    147068    NC_067196.1:147068    147068.0  0.048780   \n",
       "2         NC_067196.1    147069    NC_067196.1:147069    147069.0  0.024390   \n",
       "3         NC_067196.1    147070    NC_067196.1:147070    147070.0  0.000000   \n",
       "4         NC_067196.1    147072    NC_067196.1:147072    147072.0  0.018293   \n",
       "...               ...       ...                   ...         ...       ...   \n",
       "28585158  NC_067202.1  19477723  NC_067202.1:19477723  19477700.0  0.000000   \n",
       "28585159  NC_067202.1  19477730  NC_067202.1:19477730  19477700.0  0.036585   \n",
       "28585160  NC_067202.1  19477740  NC_067202.1:19477740  19477700.0  0.000000   \n",
       "28585161  NC_067202.1  19477770  NC_067202.1:19477770  19477800.0  0.390244   \n",
       "28585162  NC_067202.1  19477785  NC_067202.1:19477785  19477800.0  0.012195   \n",
       "\n",
       "              ihh1        p2      ihh2     xpehh  normxpehh  crit  \\\n",
       "0         3849.020  0.381356  2542.010  0.180172    1.58557     0   \n",
       "1         3270.680  0.186441  1366.650  0.378980    2.42020     1   \n",
       "2         3022.890  0.000000  1374.130  0.342395    2.26661     1   \n",
       "3         3023.680  0.000000  1374.130  0.342510    2.26709     1   \n",
       "4         3023.510  0.000000  1374.130  0.342485    2.26699     1   \n",
       "...            ...       ...       ...       ...        ...   ...   \n",
       "28585158   398.509  0.000000   270.019  0.169044    1.53885     0   \n",
       "28585159   396.811  0.000000   270.019  0.167189    1.53106     0   \n",
       "28585160   409.008  0.000000   270.019  0.180337    1.58626     0   \n",
       "28585161   333.042  0.245763   207.029  0.206467    1.69596     0   \n",
       "28585162   379.232  0.000000   236.080  0.205845    1.69335     0   \n",
       "\n",
       "          genomic_position    color  norm_p_value  p_value_bonferroni  \\\n",
       "0                   147067  #078d70      0.112837                 1.0   \n",
       "1                   147068  #078d70      0.015512                 1.0   \n",
       "2                   147069  #078d70      0.023414                 1.0   \n",
       "3                   147070  #078d70      0.023385                 1.0   \n",
       "4                   147072  #078d70      0.023391                 1.0   \n",
       "...                    ...      ...           ...                 ...   \n",
       "28585158         317910191  #078d70      0.123841                 1.0   \n",
       "28585159         317910198  #078d70      0.125755                 1.0   \n",
       "28585160         317910208  #078d70      0.112680                 1.0   \n",
       "28585161         317910238  #078d70      0.089893                 1.0   \n",
       "28585162         317910253  #078d70      0.090389                 1.0   \n",
       "\n",
       "               fdr  nw_intro  \n",
       "0         0.999997  0.823171  \n",
       "1         0.562670  0.823171  \n",
       "2         0.679088  0.823171  \n",
       "3         0.678738  0.823171  \n",
       "4         0.678825  0.823171  \n",
       "...            ...       ...  \n",
       "28585158  0.999997  0.018293  \n",
       "28585159  0.999997  0.018293  \n",
       "28585160  0.999997  0.018293  \n",
       "28585161  0.999997  0.018293  \n",
       "28585162  0.999997  0.018293  \n",
       "\n",
       "[28585163 rows x 17 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2688d3e8-43bd-4647-85d6-209f04cdf1df",
   "metadata": {},
   "outputs": [],
   "source": [
    "fst_df = pd.read_csv(f\"{results_dir}/fst/genome_wide_fsts.csv\", sep=\",\", header=0)\n",
    "\n",
    "\n",
    "rank = fst_df[\"sh_nw_v_sh_se\"].rank() / np.isfinite(fst_df[\"sh_nw_v_sh_se\"]).sum()\n",
    "sig_fst_df = fst_df.loc[(rank >= 0.95)]\n",
    "\n",
    "fst_bed_df = sig_fst_df[[\"chr\", \"start\", \"stop\"]]\n",
    "fst_bed_df.to_csv(\"fst_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "0e50135a-c385-4003-9f3c-c86a09d41d10",
   "metadata": {},
   "outputs": [],
   "source": [
    "rf_df = pd.read_csv(f\"{results_dir}/rfmix/rfmix_perc_bovis_genome.csv\", sep=\",\", header=0)\n",
    "high_anc_df = rf_df.loc[rf_df[\"perc_sh_nw\"] >= 0.95]\n",
    "\n",
    "anc_bed_df = high_anc_df[[\"chrom\", \"s_pos\", \"e_pos\"]]\n",
    "anc_bed_df.to_csv(\"ancestry.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "dacd06d4-0ece-43d5-8720-272bdc62efa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fst_df = pd.read_csv(f\"{results_dir}/fst/genome_wide_fsts.csv\", sep=\",\", header=0)\n",
    "\n",
    "\n",
    "rank = fst_df[\"sh_nw_v_sh_se\"].rank() / np.isfinite(fst_df[\"sh_nw_v_sh_se\"]).sum()\n",
    "sig_fst_df = fst_df.loc[(rank >= 0.95)]\n",
    "\n",
    "fst_bed_df = sig_fst_df[[\"chr\", \"start\", \"stop\"]]\n",
    "fst_bed_df.to_csv(\"fst_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "02b7033f-be6f-47ae-b98d-7ef1f59ee755",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dtd_df = pd.read_csv(f\"{results_dir}/delta_td/delta_tajimas_d.csv\", sep=\",\", header=0)\n",
    "\n",
    "\n",
    "# sig_dtd_df = dtd_df.loc[dtd_df[\"significant\"]]\n",
    "\n",
    "\n",
    "# dtd_bed_df = sig_dtd_df[[\"chrom\", \"start\", \"stop\"]]\n",
    "# dtd_bed_df.to_csv(\"delta_td_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0271fe35-ba67-4cff-92c8-433ed0ee8034",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_df = pd.read_csv(f\"{results_dir}/abba_baba/patterson_d.csv\", sep=\",\", header=0)\n",
    "\n",
    "\n",
    "sig_d_df = d_df.loc[d_df[\"z_score\"]>2]\n",
    "sig_d_df\n",
    "\n",
    "d_bed_df = sig_d_df[[\"chrom\", \"start\", \"stop\"]]\n",
    "d_bed_df.to_csv(\"d_outlier.bed\", sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "d2ae7395-9804-454b-9fdd-84c92a8ad058",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given data\n",
    "mu = d_df[\"d\"].std().mean()\n",
    "sigma = d_df[\"d\"].std()  # You need to provide this value\n",
    "alpha = 0.05  # For a 95% confidence level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "319ae8ea-17a9-4384-acef-b482305f248a",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'norm' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[52], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Two-tailed test critical Z-values\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m z_critical \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[43mnorm\u001b[49m\u001b[38;5;241m.\u001b[39mppf(alpha\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m2\u001b[39m)  \u001b[38;5;66;03m# for a two-tailed test\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'norm' is not defined"
     ]
    }
   ],
   "source": [
    "# Two-tailed test critical Z-values\n",
    "z_critical = -norm.ppf(alpha/2)  # for a two-tailed test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "0d1eb032-ba89-425a-97ec-aa3935afe3ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'norm' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[50], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Two-tailed test critical Z-values\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m z_critical \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[43mnorm\u001b[49m\u001b[38;5;241m.\u001b[39mppf(alpha\u001b[38;5;241m/\u001b[39m\u001b[38;5;241m2\u001b[39m)  \u001b[38;5;66;03m# for a two-tailed test\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Sample data points\u001b[39;00m\n\u001b[1;32m      5\u001b[0m data \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([\u001b[38;5;241m0.2\u001b[39m, \u001b[38;5;241m0.4\u001b[39m, \u001b[38;5;241m0.6\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.3\u001b[39m, \u001b[38;5;241m.\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;241m.\u001b[39m])  \u001b[38;5;66;03m# Fill in your data points\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'norm' is not defined"
     ]
    }
   ],
   "source": [
    "# Sample data points\n",
    "data = np.array([0.2, 0.4, 0.6, -0.3, ...])  # Fill in your data points\n",
    "\n",
    "# Calculate Z-scores for each data point\n",
    "z_scores = (data - mu) / sigma\n",
    "\n",
    "# Identify values significantly different from 0\n",
    "significant_values = data[np.abs(z_scores) > z_critical]\n",
    "print(significant_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c57b863e-f4eb-4d20-979c-ff25874bfb61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.42992946612652355"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_df[\"d\"].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "82cb8ee7-3f5a-49ab-a858-d4acf181c5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "conda run -n popstructure bedtools intersect -c -a 1p_windows.bed -b fst_outlier.bed | awk '{if ($4 > 0) print $0}' >fst_counts.bed\n",
    "conda run -n popstructure bedtools intersect -c -a 1p_windows.bed -b normxpehh_outlier.bed | awk '{if ($4 > 0) print $0}' >xpehh_counts.bed\n",
    "conda run -n popstructure bedtools intersect -c -a 1p_windows.bed -b ancestry.bed | awk '{if ($4 > 0) print $0}' >anc_counts.bed\n",
    "conda run -n popstructure bedtools intersect -c -a 1p_windows.bed -b d_outlier.bed | awk '{if ($4 > 0) print $0}' >d_counts.bed\n",
    "\n",
    "conda run -n popstructure bedtools multiinter -header -i fst_counts.bed xpehh_counts.bed anc_counts.bed d_counts.bed >test.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d085ab99-ed0f-469f-85c0-8eb2496f6c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>num</th>\n",
       "      <th>list</th>\n",
       "      <th>fst_counts.bed</th>\n",
       "      <th>xpehh_counts.bed</th>\n",
       "      <th>anc_counts.bed</th>\n",
       "      <th>d_counts.bed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>28348440</td>\n",
       "      <td>28877530</td>\n",
       "      <td>4</td>\n",
       "      <td>1,2,3,4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>9712340</td>\n",
       "      <td>10514400</td>\n",
       "      <td>4</td>\n",
       "      <td>1,2,3,4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chrom     start       end  num     list  fst_counts.bed  \\\n",
       "229  NC_067199.1  28348440  28877530    4  1,2,3,4               1   \n",
       "263  NC_067200.1   9712340  10514400    4  1,2,3,4               1   \n",
       "\n",
       "     xpehh_counts.bed  anc_counts.bed  d_counts.bed  \n",
       "229                 1               1             1  \n",
       "263                 1               1             1  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(\"test.bed\", sep=\"\\t\", header=0)\n",
    "df = df.loc[df[\"num\"] ==4]\n",
    "df[[\"chrom\", \"start\", \"end\"]].to_csv(\"target_introgressed_loci.bed\", sep=\"\\t\", header=False, index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0a5cc0e2-7648-4b9f-8d99-c4c5f990a470",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.33115"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(df[\"end\"] - df[\"start\"])/1e6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eac5d3c-6630-428d-8353-cc74fc319089",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "conda run -n popstructure \\\n",
    "    vcftools \\\n",
    "        --vcf ../filter_genotypes/sorted_annotated_snps.vcf \\\n",
    "        --recode \\\n",
    "        --recode-INFO-all \\\n",
    "        --out target_introgressed_loci \\\n",
    "        --bed target_introgressed_loci.bed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "popstructure",
   "language": "python",
   "name": "popstructure"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
