{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68cd7802-2282-44d4-bc0f-fa47b46dc088",
   "metadata": {},
   "source": [
    "### questions from TJC\n",
    " - can we get estimates/error bars from wihtin individuals when it comes to dating admixture\n",
    " - can we get individual admixture ages per country (maybe a boxplot).\n",
    " - test for selection consider using the same ones from the previous paper and/or iHS."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b68a76ed-f7ae-4980-bba1-c562dbeca49c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Ancestry *S.haematobium* vs. *S. bovis* across Africa (SCAN)\n",
    "\n",
    "Neal Platt 2021"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2259ab5f-d036-4485-a195-de28ee4cd0ee",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Prep envrinoment and data for analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1e661df-6cfd-405e-9de3-75a4178c196c",
   "metadata": {},
   "source": [
    "use the `scan_selection` conda environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c35122a-e458-4bec-9c29-c875fa386805",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import allel\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "from tqdm import tqdm\n",
    "import statsmodels.stats.multitest as smm\n",
    "import scipy.stats as stats\n",
    "from pathlib import Path\n",
    "from matplotlib.colors import ListedColormap\n",
    "# from itertools import groupby\n",
    "# import seaborn as sns\n",
    "# import pickle\n",
    "# #from collections import Counter\n",
    "# from IPython.display import Image \n",
    "# import glob\n",
    "# #import pybedtools\n",
    "# import gffutils\n",
    "# from scipy.signal import medfilt\n",
    "from collections import defaultdict\n",
    "from intervaltree import Interval, IntervalTree\n",
    "#import vcf\n",
    "#import loter.locanc.local_ancestry as lc\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cfaa4a9e-a910-44db-9415-79d3a6185878",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_scan\"\n",
    "results_dir=f\"{proj_dir}/results\"\n",
    "os.chdir(results_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1edb0330-537c-46d0-b50b-fd52f62adf06",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get sample info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a4837a2-c41e-440d-8b18-56aa40ea391d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info_df = pd.read_csv(\"{}/data/seq_and_sample_docs/all_sh_sb_sample_data.csv\".format(proj_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7da20ad0-918c-47de-88fd-c95dffa58dcc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>admera_id</th>\n",
       "      <th>barcode_label_ID</th>\n",
       "      <th>library_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>presumed_species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>host</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>project</th>\n",
       "      <th>data_type</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19047FL-14-02-03</td>\n",
       "      <td>MCF03050E0612</td>\n",
       "      <td>na</td>\n",
       "      <td>sha_angola_cota_57</td>\n",
       "      <td>SRR11907442</td>\n",
       "      <td>s_sp</td>\n",
       "      <td>angola</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-9.26186</td>\n",
       "      <td>16.13693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19047FL-14-02-04</td>\n",
       "      <td>MCF03050E0613</td>\n",
       "      <td>na</td>\n",
       "      <td>sha_angola_cota_58</td>\n",
       "      <td>SRR11907441</td>\n",
       "      <td>s_sp</td>\n",
       "      <td>angola</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-9.26186</td>\n",
       "      <td>16.13693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19047FL-14-02-05</td>\n",
       "      <td>MCF03050E0614</td>\n",
       "      <td>na</td>\n",
       "      <td>sha_angola_cota_59</td>\n",
       "      <td>SRR11907440</td>\n",
       "      <td>s_sp</td>\n",
       "      <td>angola</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-9.26186</td>\n",
       "      <td>16.13693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19047FL-14-02-06</td>\n",
       "      <td>MCF03050E0615</td>\n",
       "      <td>na</td>\n",
       "      <td>sha_angola_icau_60</td>\n",
       "      <td>SRR11907439</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>angola</td>\n",
       "      <td>na</td>\n",
       "      <td>human</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-8.64510</td>\n",
       "      <td>13.79642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19047FL-14-02-07</td>\n",
       "      <td>MCF03050E0615</td>\n",
       "      <td>na</td>\n",
       "      <td>sha_angola_icau_61</td>\n",
       "      <td>SRR11907438</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>angola</td>\n",
       "      <td>na</td>\n",
       "      <td>human</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>-8.64510</td>\n",
       "      <td>13.79642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>sh_egypt_SRR433862</td>\n",
       "      <td>SRR433862</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>egypt</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>Rey_2019</td>\n",
       "      <td>genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>sh_gambia_SRR13579872</td>\n",
       "      <td>SRR13579872</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>gambia</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>Rey_2019</td>\n",
       "      <td>genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>sh_ivorycoast_SRR13579866</td>\n",
       "      <td>SRR13579866</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>cote_d_ivoire</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>Rey_2019</td>\n",
       "      <td>genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>sh_madagascar_SRR13579871</td>\n",
       "      <td>SRR13579871</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>madagascar</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>Rey_2019</td>\n",
       "      <td>genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>sh_nigeria_SRR13579882</td>\n",
       "      <td>SRR13579882</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>haematobium</td>\n",
       "      <td>Rey_2019</td>\n",
       "      <td>genome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>449 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            admera_id barcode_label_ID library_id                sample_name  \\\n",
       "0    19047FL-14-02-03    MCF03050E0612         na         sha_angola_cota_57   \n",
       "1    19047FL-14-02-04    MCF03050E0613         na         sha_angola_cota_58   \n",
       "2    19047FL-14-02-05    MCF03050E0614         na         sha_angola_cota_59   \n",
       "3    19047FL-14-02-06    MCF03050E0615         na         sha_angola_icau_60   \n",
       "4    19047FL-14-02-07    MCF03050E0615         na         sha_angola_icau_61   \n",
       "..                ...              ...        ...                        ...   \n",
       "444                na               na         na         sh_egypt_SRR433862   \n",
       "445                na               na         na      sh_gambia_SRR13579872   \n",
       "446                na               na         na  sh_ivorycoast_SRR13579866   \n",
       "447                na               na         na  sh_madagascar_SRR13579871   \n",
       "448                na               na         na     sh_nigeria_SRR13579882   \n",
       "\n",
       "    sra_accession presumed_species        country locale   host    phenotype  \\\n",
       "0     SRR11907442             s_sp         angola     na  snail          unk   \n",
       "1     SRR11907441             s_sp         angola     na  snail          unk   \n",
       "2     SRR11907440             s_sp         angola     na  snail          unk   \n",
       "3     SRR11907439    s_haematobium         angola     na  human  haematobium   \n",
       "4     SRR11907438    s_haematobium         angola     na  human  haematobium   \n",
       "..            ...              ...            ...    ...    ...          ...   \n",
       "444     SRR433862    s_haematobium          egypt     na     na  haematobium   \n",
       "445   SRR13579872    s_haematobium         gambia     na     na  haematobium   \n",
       "446   SRR13579866    s_haematobium  cote_d_ivoire     na     na  haematobium   \n",
       "447   SRR13579871    s_haematobium     madagascar     na     na  haematobium   \n",
       "448   SRR13579882    s_haematobium        nigeria     na     na  haematobium   \n",
       "\n",
       "         project data_type      lat       lon  \n",
       "0    scan_africa    genome -9.26186  16.13693  \n",
       "1    scan_africa    genome -9.26186  16.13693  \n",
       "2    scan_africa    genome -9.26186  16.13693  \n",
       "3    scan_africa    genome -8.64510  13.79642  \n",
       "4    scan_africa    genome -8.64510  13.79642  \n",
       "..           ...       ...      ...       ...  \n",
       "444     Rey_2019    genome      NaN       NaN  \n",
       "445     Rey_2019    genome      NaN       NaN  \n",
       "446     Rey_2019    genome      NaN       NaN  \n",
       "447     Rey_2019    genome      NaN       NaN  \n",
       "448     Rey_2019    genome      NaN       NaN  \n",
       "\n",
       "[449 rows x 14 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f868233f-0b80-42e2-8967-ab1a6c3de57b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Read in data and get allele counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "52395106-bca1-404a-a117-b321c10ea404",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#read in the vcf file\n",
    "filtered_callset=allel.read_vcf(f'{results_dir}/post_phase_filtering/autosomes_unrelated.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "ae5d96a9-1729-4358-bcb0-05d7e3c696c5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#get relevant info\n",
    "gt      = allel.GenotypeArray(filtered_callset['calldata/GT'])\n",
    "chroms  = filtered_callset[\"variants/CHROM\"]\n",
    "samples = filtered_callset[\"samples\"]\n",
    "poss    = filtered_callset[\"variants/POS\"]\n",
    "var_ids = filtered_callset[\"variants/ID\"]\n",
    "\n",
    "uniq_chroms=np.unique(chroms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "0e2a9d1e-4401-4f4e-8626-da84259fb7e8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#read in pca df since it has species ids and kmeans labels\n",
    "pca_df=pd.read_csv(\"{}/pca/pca_df.csv\".format(results_dir), sep=\",\")\n",
    "pca_df[\"country\"] = pca_df[\"country\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "f95f20f9-e105-4f51-9526-0701d8ea5fa1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>locale</th>\n",
       "      <th>host</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>project</th>\n",
       "      <th>data_type</th>\n",
       "      <th>lat</th>\n",
       "      <th>lon</th>\n",
       "      <th>kmeans_group</th>\n",
       "      <th>pca_species</th>\n",
       "      <th>kmeans_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>sbo_cdivoire_raviart_108</td>\n",
       "      <td>0.021400</td>\n",
       "      <td>-0.070582</td>\n",
       "      <td>-0.045198</td>\n",
       "      <td>0.206487</td>\n",
       "      <td>0.004450</td>\n",
       "      <td>0.012315</td>\n",
       "      <td>0.028276</td>\n",
       "      <td>-0.103222</td>\n",
       "      <td>-0.100459</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.400372</td>\n",
       "      <td>-4.896099</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>sbo_cdivoire_raviart_109</td>\n",
       "      <td>0.020538</td>\n",
       "      <td>-0.076418</td>\n",
       "      <td>-0.046554</td>\n",
       "      <td>0.226617</td>\n",
       "      <td>-0.003926</td>\n",
       "      <td>-0.014860</td>\n",
       "      <td>-0.011537</td>\n",
       "      <td>-0.087275</td>\n",
       "      <td>-0.110759</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>7.400372</td>\n",
       "      <td>-4.896099</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>sbo_niger_libore_154</td>\n",
       "      <td>0.025617</td>\n",
       "      <td>-0.079330</td>\n",
       "      <td>-0.013864</td>\n",
       "      <td>-0.055163</td>\n",
       "      <td>-0.001331</td>\n",
       "      <td>-0.003984</td>\n",
       "      <td>0.000089</td>\n",
       "      <td>-0.016355</td>\n",
       "      <td>0.055014</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>sbo_niger_libore_155</td>\n",
       "      <td>0.021715</td>\n",
       "      <td>-0.077734</td>\n",
       "      <td>-0.013288</td>\n",
       "      <td>-0.061101</td>\n",
       "      <td>0.008696</td>\n",
       "      <td>-0.056704</td>\n",
       "      <td>0.051375</td>\n",
       "      <td>-0.009732</td>\n",
       "      <td>0.050136</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>sbo_niger_libore_160</td>\n",
       "      <td>0.021868</td>\n",
       "      <td>-0.075221</td>\n",
       "      <td>-0.018740</td>\n",
       "      <td>-0.071181</td>\n",
       "      <td>-0.001439</td>\n",
       "      <td>-0.055054</td>\n",
       "      <td>0.051839</td>\n",
       "      <td>-0.075687</td>\n",
       "      <td>0.004744</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.388050</td>\n",
       "      <td>2.199130</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>165</th>\n",
       "      <td>ssp_niger_libore_166</td>\n",
       "      <td>0.022455</td>\n",
       "      <td>-0.078463</td>\n",
       "      <td>-0.021334</td>\n",
       "      <td>-0.066131</td>\n",
       "      <td>0.005504</td>\n",
       "      <td>-0.046551</td>\n",
       "      <td>0.052623</td>\n",
       "      <td>-0.059921</td>\n",
       "      <td>0.043647</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.388050</td>\n",
       "      <td>2.199130</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>ssp_niger_libore_167</td>\n",
       "      <td>0.023576</td>\n",
       "      <td>-0.078099</td>\n",
       "      <td>-0.011150</td>\n",
       "      <td>-0.061000</td>\n",
       "      <td>0.017043</td>\n",
       "      <td>-0.045486</td>\n",
       "      <td>0.040388</td>\n",
       "      <td>-0.017674</td>\n",
       "      <td>0.024366</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>ssp_niger_libore_168</td>\n",
       "      <td>0.022321</td>\n",
       "      <td>-0.076938</td>\n",
       "      <td>-0.018901</td>\n",
       "      <td>-0.060000</td>\n",
       "      <td>0.006243</td>\n",
       "      <td>-0.004137</td>\n",
       "      <td>0.029055</td>\n",
       "      <td>-0.042276</td>\n",
       "      <td>0.080228</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.390080</td>\n",
       "      <td>2.197110</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>ssp_niger_libore_169</td>\n",
       "      <td>0.023969</td>\n",
       "      <td>-0.073719</td>\n",
       "      <td>-0.019718</td>\n",
       "      <td>-0.050199</td>\n",
       "      <td>0.016409</td>\n",
       "      <td>-0.015895</td>\n",
       "      <td>0.038597</td>\n",
       "      <td>-0.020768</td>\n",
       "      <td>0.037634</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>ssp_niger_libore_173</td>\n",
       "      <td>0.026690</td>\n",
       "      <td>-0.083394</td>\n",
       "      <td>-0.003446</td>\n",
       "      <td>-0.070261</td>\n",
       "      <td>-0.010403</td>\n",
       "      <td>-0.085303</td>\n",
       "      <td>0.096133</td>\n",
       "      <td>-0.115088</td>\n",
       "      <td>0.082561</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>snail</td>\n",
       "      <td>unk</td>\n",
       "      <td>scan_africa</td>\n",
       "      <td>genome</td>\n",
       "      <td>13.404170</td>\n",
       "      <td>2.188020</td>\n",
       "      <td>0</td>\n",
       "      <td>s_haematobium</td>\n",
       "      <td>sh_nw</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample_name         1         2         3         4  \\\n",
       "29   sbo_cdivoire_raviart_108  0.021400 -0.070582 -0.045198  0.206487   \n",
       "30   sbo_cdivoire_raviart_109  0.020538 -0.076418 -0.046554  0.226617   \n",
       "31       sbo_niger_libore_154  0.025617 -0.079330 -0.013864 -0.055163   \n",
       "32       sbo_niger_libore_155  0.021715 -0.077734 -0.013288 -0.061101   \n",
       "33       sbo_niger_libore_160  0.021868 -0.075221 -0.018740 -0.071181   \n",
       "..                        ...       ...       ...       ...       ...   \n",
       "165      ssp_niger_libore_166  0.022455 -0.078463 -0.021334 -0.066131   \n",
       "166      ssp_niger_libore_167  0.023576 -0.078099 -0.011150 -0.061000   \n",
       "167      ssp_niger_libore_168  0.022321 -0.076938 -0.018901 -0.060000   \n",
       "168      ssp_niger_libore_169  0.023969 -0.073719 -0.019718 -0.050199   \n",
       "169      ssp_niger_libore_173  0.026690 -0.083394 -0.003446 -0.070261   \n",
       "\n",
       "            5         6         7         8         9  ...  locale   host  \\\n",
       "29   0.004450  0.012315  0.028276 -0.103222 -0.100459  ...      na  snail   \n",
       "30  -0.003926 -0.014860 -0.011537 -0.087275 -0.110759  ...      na  snail   \n",
       "31  -0.001331 -0.003984  0.000089 -0.016355  0.055014  ...      na  snail   \n",
       "32   0.008696 -0.056704  0.051375 -0.009732  0.050136  ...      na  snail   \n",
       "33  -0.001439 -0.055054  0.051839 -0.075687  0.004744  ...      na  snail   \n",
       "..        ...       ...       ...       ...       ...  ...     ...    ...   \n",
       "165  0.005504 -0.046551  0.052623 -0.059921  0.043647  ...      na  snail   \n",
       "166  0.017043 -0.045486  0.040388 -0.017674  0.024366  ...      na  snail   \n",
       "167  0.006243 -0.004137  0.029055 -0.042276  0.080228  ...      na  snail   \n",
       "168  0.016409 -0.015895  0.038597 -0.020768  0.037634  ...      na  snail   \n",
       "169 -0.010403 -0.085303  0.096133 -0.115088  0.082561  ...      na  snail   \n",
       "\n",
       "     phenotype      project  data_type        lat       lon  kmeans_group  \\\n",
       "29         unk  scan_africa     genome   7.400372 -4.896099             0   \n",
       "30         unk  scan_africa     genome   7.400372 -4.896099             0   \n",
       "31         unk  scan_africa     genome  13.390080  2.197110             0   \n",
       "32         unk  scan_africa     genome  13.390080  2.197110             0   \n",
       "33         unk  scan_africa     genome  13.388050  2.199130             0   \n",
       "..         ...          ...        ...        ...       ...           ...   \n",
       "165        unk  scan_africa     genome  13.388050  2.199130             0   \n",
       "166        unk  scan_africa     genome  13.390080  2.197110             0   \n",
       "167        unk  scan_africa     genome  13.390080  2.197110             0   \n",
       "168        unk  scan_africa     genome  13.404170  2.188020             0   \n",
       "169        unk  scan_africa     genome  13.404170  2.188020             0   \n",
       "\n",
       "       pca_species  kmeans_label  \n",
       "29   s_haematobium         sh_nw  \n",
       "30   s_haematobium         sh_nw  \n",
       "31   s_haematobium         sh_nw  \n",
       "32   s_haematobium         sh_nw  \n",
       "33   s_haematobium         sh_nw  \n",
       "..             ...           ...  \n",
       "165  s_haematobium         sh_nw  \n",
       "166  s_haematobium         sh_nw  \n",
       "167  s_haematobium         sh_nw  \n",
       "168  s_haematobium         sh_nw  \n",
       "169  s_haematobium         sh_nw  \n",
       "\n",
       "[63 rows x 37 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pca_df.loc[ (pca_df[\"kmeans_label\"] == \"sh_nw\") & \n",
    "            (pca_df[\"project\"] == \"scan_africa\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a331157a-c24d-4f9e-93a5-9a6b6c572675",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "617f86c1-f995-4ccd-a088-fef5079408f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|                                                                                                                                                                                                                  | 0/2 [00:00<?, ?it/s]\u001b[A\n",
      " 50%|█████████████████████████████████████████████████████████████████████████████████████████████████████                                                                                                     | 1/2 [00:05<00:05,  5.50s/it]\u001b[A\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:20<00:00, 10.25s/it]\u001b[A\n"
     ]
    }
   ],
   "source": [
    "#get species idxs\n",
    "sp_idxs = {}\n",
    "sp_acs  = {}\n",
    "\n",
    "for sp in tqdm([\"s_bovis\", \"s_haematobium\"]):\n",
    "    \n",
    "    #find indecies for each sample from that species\n",
    "    names = pca_df.loc[pca_df[\"pca_species\"] ==  sp, \"sample_name\"]\n",
    "    sp_idxs[sp]=np.in1d(samples, names).nonzero()[0]\n",
    "\n",
    "    #get allele counts for each species\n",
    "    sp_acs[sp] = gt.count_alleles(subpop=sp_idxs[sp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "802fdee0-bee5-47f0-8701-ca86bae3da30",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add outgroups\n",
    "s_guineensis   = \"guineensis_saotome_ERR119612\"\n",
    "s_intercalatum = \"intercalatum_drcongo_ERR119613\"\n",
    "s_margrebowiei = \"margrebowiei_zambia_ERR310940\"\n",
    "s_matthei      = \"matthei_zambia_ERR103051\"\n",
    "\n",
    "outgroups=[s_guineensis, s_intercalatum, s_margrebowiei, s_matthei]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "0844b074-df94-4c1c-adf9-812367469348",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|                                                                                                                                                                                                                  | 0/4 [00:00<?, ?it/s]\u001b[A\n",
      " 25%|██████████████████████████████████████████████████▌                                                                                                                                                       | 1/4 [00:02<00:07,  2.52s/it]\u001b[A\n",
      " 50%|█████████████████████████████████████████████████████████████████████████████████████████████████████                                                                                                     | 2/4 [00:04<00:04,  2.36s/it]\u001b[A\n",
      " 75%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌                                                  | 3/4 [00:06<00:02,  2.29s/it]\u001b[A\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4/4 [00:08<00:00,  2.23s/it]\u001b[A\n"
     ]
    }
   ],
   "source": [
    "for sp in tqdm(outgroups):\n",
    "    sp_idxs[sp] = [list(samples).index(sp)]\n",
    "    sp_acs[sp] = gt.count_alleles(subpop=sp_idxs[sp])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e258bead-9b7e-4ed7-98b3-01d65e783eb6",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Kmeans group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "82be9c18-5dab-4a8b-bc11-2c1bf233a5f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  0%|                                                                                                                                                                                                                  | 0/7 [00:00<?, ?it/s]\u001b[A\n",
      " 14%|████████████████████████████▊                                                                                                                                                                             | 1/7 [00:03<00:20,  3.44s/it]\u001b[A\n",
      " 29%|█████████████████████████████████████████████████████████▋                                                                                                                                                | 2/7 [00:12<00:33,  6.75s/it]\u001b[A\n",
      " 43%|██████████████████████████████████████████████████████████████████████████████████████▌                                                                                                                   | 3/7 [00:14<00:18,  4.57s/it]\u001b[A\n",
      " 57%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████▍                                                                                      | 4/7 [00:16<00:10,  3.54s/it]\u001b[A\n",
      " 71%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                                         | 5/7 [00:18<00:05,  2.98s/it]\u001b[A\n",
      " 86%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                            | 6/7 [00:20<00:02,  2.64s/it]\u001b[A\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [00:27<00:00,  3.90s/it]\u001b[A\n"
     ]
    }
   ],
   "source": [
    "#get species idxs\n",
    "k_pop_idxs = {}\n",
    "k_pop_acs  = {}\n",
    "\n",
    "for k in tqdm(pca_df[\"kmeans_label\"].unique()):\n",
    "    \n",
    "    #find indecies for each sample per k population\n",
    "    ks = pca_df.loc[pca_df[\"kmeans_label\"] ==  k, \"sample_name\"]\n",
    "    k_pop_idxs[k]=np.in1d(samples, ks).nonzero()[0]\n",
    "    \n",
    "    #get allele counts for k group\n",
    "    k_pop_acs[k] = gt.count_alleles(subpop=k_pop_idxs[k])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "9156c16c-ab11-4f7d-a252-029b1e57860f",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_acs={\"sh_nw\":          k_pop_acs[\"sh_nw\"],\n",
    "           \"sh_se\":          k_pop_acs[\"sh_se\"],\n",
    "           \"sb\":             k_pop_acs[\"sb\"] ,\n",
    "           \"sh\":             sp_acs[\"s_haematobium\"],\n",
    "           \"s_guineensis\":   sp_acs[\"guineensis_saotome_ERR119612\"],\n",
    "           \"s_intercalatum\": sp_acs[\"intercalatum_drcongo_ERR119613\"],\n",
    "           \"s_margrebowiei\": sp_acs[\"margrebowiei_zambia_ERR310940\"],\n",
    "           \"s_matthei\":      sp_acs[\"matthei_zambia_ERR103051\"]  }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a229d02-1dad-48e6-b2c5-99ad4dae74b3",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Selscan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c7bc246-9d1d-4880-a5dd-2ebb4d627164",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(\"{}/selscan\".format(results_dir)).mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(\"{}/selscan\".format(results_dir))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9588ed91-8b85-4141-ad91-8ea7a93b79a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ln -s ../post_phase_filtering/chrs_unrelated.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "225c925f-0557-4106-86b9-ee497deb6b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#samples with no lat/lon info cause major problems will need to exclude downstream\n",
    "sh_nw_samples = pca_df.loc[pca_df[\"kmeans_label\"] == \"sh_nw\", \"sample_name\" ].values\n",
    "sh_se_samples = pca_df.loc[pca_df[\"kmeans_label\"] == \"sh_se\", \"sample_name\" ].values\n",
    "sb_samples    = pca_df.loc[pca_df[\"kmeans_label\"] == \"sb\"   , \"sample_name\" ].values\n",
    "\n",
    "np.savetxt(\"nw.list\", sh_nw_samples, delimiter=',', fmt=\"%s\")\n",
    "np.savetxt(\"se.list\", sh_se_samples, delimiter=',', fmt=\"%s\")\n",
    "np.savetxt(\"sb.list\", sb_samples, delimiter=',', fmt=\"%s\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b8b898-45e1-405c-ba73-793333cd69e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "vcftools \\\n",
    "    --vcf chrs_unrelated.vcf \\\n",
    "    --keep nw.list \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >nw.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "769cc842-8ed6-4ee6-b5b9-529a4a07363e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "vcftools \\\n",
    "    --vcf chrs_unrelated.vcf \\\n",
    "    --keep se.list \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >se.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6bc9bf8-de88-4e97-923d-82ce11c0fee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "vcftools \\\n",
    "    --vcf chrs_unrelated.vcf \\\n",
    "    --keep sb.list \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >sb.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd144116-c9e8-4343-8836-fadf7391fe54",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat nw.list se.list >sh.list\n",
    "\n",
    "vcftools \\\n",
    "    --vcf chrs_unrelated.vcf \\\n",
    "    --keep sh.list \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --stdout \\\n",
    "    >sh.vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12396389-7b90-4532-bd3b-ad64cc32519e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create a map file\n",
    "#< chr # > <id > < genetic position > < physical position >\n",
    "\n",
    " cut -f1-3 chrs_unrelated.vcf | grep -v \"#\" | awk '{print $1\"\\t\"$3\"\\t\"$2/287000\"\\t\"$2}' >selscan.map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da57009a-5ecf-4f94-aec4-1b407cf3e731",
   "metadata": {},
   "outputs": [],
   "source": [
    "for CHROM in \"NC_067195.1\" \"NC_067196.1\" \"NC_067197.1\" \"NC_067198.1\" \"NC_067199.1\" \"NC_067200.1\" \"NC_067201.1\" \"NC_067202.1\"; do\n",
    "    for POP in \"nw\" \"sh\" \"se\" \"sb\"; do\n",
    "        conda run -n scan_selection --cwd . --live-stream vcftools --vcf ${POP}.vcf --chr ${CHROM} --recode --recode-INFO-all --stdout >${POP}_${CHROM}.vcf &\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383f1ed9-b6ef-4327-88e9-7bfaeaa6a22f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#./selscan/bin/linux/selscan --xpehh --unphased --vcf nw_${CHROM}.vcf  --vcf-ref se_${CHROM}.vcf --map selscan.map --out nw_vs_se_ref_${CHROM} --threads 12 --pmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ee3d988-105e-41cd-9ede-c22797af4a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY=\"nw\"\n",
    "REF=\"se\"\n",
    "\n",
    "for CHROM in \"NC_067195.1\" \"NC_067196.1\" \"NC_067197.1\" \"NC_067198.1\" \"NC_067199.1\" \"NC_067200.1\" \"NC_067201.1\" \"NC_067202.1\"; do\n",
    "    \n",
    "    JID=\"${QUERY}-query_vs_${REF}-ref_${CHROM}_unphased\"\n",
    "    VCF=\"${QUERY}_${CHROM}.vcf\"\n",
    "    REF_VCF=\"${REF}_${CHROM}.vcf\"\n",
    "    THREADS=\"12\"\n",
    "    \n",
    "    CMD=\"conda run -n scan_selection --cwd . --live-stream ./selscan/bin/linux/selscan --xpehh --unphased --vcf ${VCF} --vcf-ref ${REF_VCF} --map selscan.map --out ${JID} --threads ${THREADS} --pmap\"\n",
    "    \n",
    "    QSUB=\"qsub -V -cwd -S /bin/bash -q all.q -j y -pe smp ${THREADS} -N ${JID} -o ${JID}.sge.log\"\n",
    "       \n",
    "    echo $CMD | $QSUB\n",
    "\n",
    "done\n",
    "\n",
    "#### THIS TOOK SEVERAL DAYS TO RUN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb085880-43ad-4d8b-8c36-e50c8e9a18a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "QUERY=\"nw\"\n",
    "REF=\"se\"\n",
    "\n",
    "for CHROM in \"NC_067195.1\"; do\n",
    "    \n",
    "    JID=\"${QUERY}-query_vs_${REF}-ref_${CHROM}_unphased\"\n",
    "    VCF=\"${QUERY}_${CHROM}.vcf\"\n",
    "    REF_VCF=\"${REF}_${CHROM}.vcf\"\n",
    "    THREADS=\"12\"\n",
    "    \n",
    "    CMD=\"conda run -n scan_selection --cwd . --live-stream ./selscan/bin/linux/selscan --xpehh --unphased --vcf ${VCF} --vcf-ref ${REF_VCF} --map selscan.map --out ${JID} --threads ${THREADS} --pmap\"\n",
    "    \n",
    "    QSUB=\"qsub -V -cwd -S /bin/bash -q all.q -j y -pe smp ${THREADS} -N ${JID} -o ${JID}.sge.log\"\n",
    "       \n",
    "    echo $CMD | $QSUB\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "16e8d5da-263c-47d5-b91f-800dea7a0253",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniq_chroms= [\"NC_067195.1\", \"NC_067196.1\", \"NC_067197.1\", \"NC_067198.1\", \"NC_067199.1\", \"NC_067200.1\", \"NC_067201.1\", \"NC_067202.1\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3ce9b9a6-4873-4e51-a11e-4810c0f41dac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:49<00:00,  6.19s/it]\n"
     ]
    }
   ],
   "source": [
    "sel_df=pd.DataFrame()\n",
    "for chrom in tqdm(uniq_chroms):\n",
    "    tmp_df=pd.read_csv(f\"nw-query_vs_se-ref_{chrom}_unphased.xpehh.out.norm\", sep=\"\\t\")\n",
    "    sel_df=pd.concat([sel_df, tmp_df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7c4261d7-be54-4d55-a80d-e2f9076f0128",
   "metadata": {},
   "outputs": [],
   "source": [
    "chroms = sel_df['id'].str.split(':').str[0]\n",
    "\n",
    "sel_df[\"chr\"] = chroms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "034ec521-eb70-492a-8a54-e6c732d44e42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NC_067195.1', 'NC_067196.1', 'NC_067197.1', 'NC_067198.1',\n",
       "       'NC_067199.1', 'NC_067200.1', 'NC_067201.1', 'NC_067202.1'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_df[\"chr\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "abd6d182-bc7b-460e-ad3e-32fe95f5a65d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_df = sel_df[[\"chr\", \"pos\",  \"id\", \"gpos\", \"p1\", \"ihh1\", \"p2\", \"ihh2\", \"xpehh\", \"normxpehh\", \"crit\"]]\n",
    "sel_df.to_csv(\"sel_df_norm.csv\", sep=\",\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a570efff-d6e1-47cb-a7c5-ebe95e17e333",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEDCAYAAAAlRP8qAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAAsTAAALEwEAmpwYAAAP20lEQVR4nO3df6jdd33H8edrbSpsLXSQuHZp2rgRHaug1lDblY1s6ugvLBudVIZuKosWHRZ0zClUmf/UjQlq1JDNIoVa3bDW4FJ/DITWsZQmIf0Zu0WN9K6ZjVUbQwsu7L0/zjf19PScnHNzz73n3E+eDzjknO/nc8958zm3r37u53zP55uqQpK0+v3SrAuQJE2HgS5JjTDQJakRBrokNcJAl6RGGOiS1IiZBnqSW5M8meThCfu/McmjSR5J8vnlrk+SVpPM8jz0JL8HHANuq6qXj+m7Cfhn4A+q6idJXlxVT65EnZK0Gsx0hl5V9wA/7j+W5DeTfC3J3iT3JvmtrukvgE9V1U+6nzXMJanPPK6h7wD+sqpeDbwP+HR3/KXAS5P8e5LdSa6cWYWSNIfOnHUB/ZKcDfwO8C9JThx+UffvmcAmYAtwAXBvkpdX1U9XuExJmktzFej0/mL4aVW9ckjbArC7qv4X+H6Sx+gF/P0rWJ8kza25WnKpqqP0wvpPANLziq75LuD3u+Nr6S3BfG8WdUrSPJr1aYt3AP8BvCzJQpK3A38KvD3JA8AjwHVd968DTyV5FPgW8FdV9dQs6pakeTTT0xYlSdMzV0sukqRTN7MPRdeuXVsbN26c1ctL0qq0d+/eH1XVumFtMwv0jRs3smfPnlm9vCStSkl+MKrNJRdJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrEvO2HLs3Mxvf/63P3D91yzQwrkU6NM3RJaoSBLkmNMNAlqRGuoeu01r9uPuq46+laLZyhS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNWLsbotJNgC3AecB/wfsqKqPD/TZAnwF+H536M6q+tupVirNiDsvarWYZPvc48B7q2pfknOAvUm+WVWPDvS7t6qunX6JkqRJjA30qjoMHO7u/yzJAWA9MBjo0twbtf+51IJFraEn2Qi8CrhvSPPlSR5IcneSi0f8/NYke5LsOXLkyOKrlSSNNHGgJzkb+BJwU1UdHWjeB1xUVa8APgncNew5qmpHVW2uqs3r1q07xZIlScNMFOhJ1tAL89ur6s7B9qo6WlXHuvu7gDVJ1k61UknSSY0N9CQBPgscqKqPjehzXtePJJd2z/vUNAuVJJ3cJGe5XAG8GXgoyf7u2AeACwGqajtwPXBjkuPAs8ANVVXTL1eSNMokZ7l8G8iYPtuAbdMqSpK0eH5TVJIaYaBLUiMMdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJasQkm3NJq5pXKdLpwkCXFsELRmueueQiSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSI8YGepINSb6V5ECSR5K8Z0ifJPlEkoNJHkxyyfKUK0kaZZILXBwH3ltV+5KcA+xN8s2qerSvz1XApu72GuAz3b+SpBUydoZeVYeral93/2fAAWD9QLfrgNuqZzdwbpLzp16tJGmkRa2hJ9kIvAq4b6BpPfB43+MFXhj6JNmaZE+SPUeOHFlkqZKkk5n4mqJJzga+BNxUVUcHm4f8SL3gQNUOYAfA5s2bX9AuTYsXhtbpaKIZepI19ML89qq6c0iXBWBD3+MLgCeWXp4kaVKTnOUS4LPAgar62IhuO4G3dGe7XAY8XVWHp1inJGmMSZZcrgDeDDyUZH937APAhQBVtR3YBVwNHASeAd469UolSSc1NtCr6tsMXyPv71PAu6ZVlCRp8fymqCQ1wkCXpEZMfNqipOfrPzXy0C3XzLASqccZuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaYaBLUiMMdElqhNvnqhn929lKpyNn6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGGOiS1AgDXZIaMTbQk9ya5MkkD49o35Lk6ST7u9vN0y9TkjTOJHu5fA7YBtx2kj73VtW1U6lIWoX695E5dMs1M6xEp7OxM/Squgf48QrUIklagmmtoV+e5IEkdye5eFSnJFuT7Emy58iRI1N6aUkSTCfQ9wEXVdUrgE8Cd43qWFU7qmpzVW1et27dFF5aknTCkgO9qo5W1bHu/i5gTZK1S65MkrQoSw70JOclSXf/0u45n1rq80qSFmfsWS5J7gC2AGuTLAAfAtYAVNV24HrgxiTHgWeBG6qqlq1iSdJQYwO9qt40pn0bvdMaJUkz5DdFJakRXiRaq5oXhpZ+wRm6JDXCQJekRhjoktQIA12SGmGgS1IjDHRJaoSBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjTDQJakRBrokNcJAl6RGeAk6acr6L4t36JZrZliJTjcGulYdryMqDeeSiyQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWrE2EBPcmuSJ5M8PKI9ST6R5GCSB5NcMv0yJUnjTDJD/xxw5UnarwI2dbetwGeWXpYkabHGBnpV3QP8+CRdrgNuq57dwLlJzp9WgZKkyUxjDX098Hjf44XumCRpBU0j0DPkWA3tmGxNsifJniNHjkzhpSVJJ0wj0BeADX2PLwCeGNaxqnZU1eaq2rxu3bopvLQk6YRpBPpO4C3d2S6XAU9X1eEpPK8kaRHG7raY5A5gC7A2yQLwIWANQFVtB3YBVwMHgWeAty5XsZKk0cYGelW9aUx7Ae+aWkWSpFPiN0UlqREGuiQ1wisWaVXwKkXSeM7QJakRBrokNcJAl6RGuIYuLaP+tf9Dt1wzw0p0OnCGLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhphoEtSIwx0SWqEgS5JjfCboppb7rAoLY4zdElqhIEuSY0w0CWpEQa6JDXCQJekRhjoktQIA12SGuF56NIK8epFWm7O0CWpEQa6JDXCJRfNFb/uL526iWboSa5M8liSg0neP6R9S5Knk+zvbjdPv1RJ0smMnaEnOQP4FPB6YAG4P8nOqnp0oOu9VXXtMtQoSZrAJDP0S4GDVfW9qvo58AXguuUtS5K0WJME+nrg8b7HC92xQZcneSDJ3Ukunkp1kqSJTfKhaIYcq4HH+4CLqupYkquBu4BNL3iiZCuwFeDCCy9cXKWSpJOaZIa+AGzoe3wB8ER/h6o6WlXHuvu7gDVJ1g4+UVXtqKrNVbV53bp1SyhbkjRokkC/H9iU5CVJzgJuAHb2d0hyXpJ09y/tnvepaRcrSRpt7JJLVR1P8m7g68AZwK1V9UiSd3bt24HrgRuTHAeeBW6oqsFlGUnSMproi0XdMsqugWPb++5vA7ZNtzSdLvwykTQdflNUmgE36tJycC8XSWqEgS5JjTDQJakRBrokNcJAl6RGeJaLZsJTFaXpc4YuSY0w0CWpES65SDPml4w0Lc7QJakRBrokNcIlF60Yz2yRlpczdElqhIEuSY0w0CWpEa6hS3PEUxi1FAa6lpUfhEorxyUXSWqEgS5JjXDJRVPnMst0uJ6uxXKGLkmNMNAlqREuuWgqXGaRZs9Al1YB19M1CQNdp8xZuTRfDHQtiiE+e87WNYqBrrEM8flluKufga6hDPHVx3DXRIGe5Erg48AZwD9V1S0D7enarwaeAf68qvZNuVYtA4O7TYb76WlsoCc5A/gU8HpgAbg/yc6qerSv21XApu72GuAz3b+aEwb36WvUe2/Qt2eSGfqlwMGq+h5Aki8A1wH9gX4dcFtVFbA7yblJzq+qw1OvWIazpmKpv0f+D2H+TBLo64HH+x4v8MLZ97A+64HnBXqSrcDW7uGxJI8tqtqlWQv8aAVfbxpWY82wOuu25kXKR0/pxxznpbtoVMMkgZ4hx+oU+lBVO4AdE7zm1CXZU1WbZ/Hap2o11gyrs25rXhnWvLwm2ctlAdjQ9/gC4IlT6CNJWkaTBPr9wKYkL0lyFnADsHOgz07gLem5DHja9XNJWlljl1yq6niSdwNfp3fa4q1V9UiSd3bt24Fd9E5ZPEjvtMW3Ll/Jp2wmSz1LtBprhtVZtzWvDGteRumdmCJJWu3cD12SGmGgS1Ijmg30JF9Msr+7HUqyf0S/Q0ke6vrtWeEyB2v5cJL/7qv76hH9rkzyWJKDSd6/0nUO1PL3Sb6T5MEkX05y7oh+Mx/ncePWfaj/ia79wSSXzKLOvno2JPlWkgNJHknyniF9tiR5uu935uZZ1Dpo3Ps9h2P9sr4x3J/kaJKbBvrM5Vg/T1U1fwP+Abh5RNshYO2sa+xq+TDwvjF9zgC+C/wGcBbwAPDbM6z5D4Ezu/sfBT46j+M8ybjR+2D/bnrfq7gMuG/Gvw/nA5d0988B/nNIzVuAr86yzlN5v+dtrIf8rvwPcNFqGOv+W7Mz9BO6jcPeCNwx61qm5LmtGKrq58CJrRhmoqq+UVXHu4e76X0HYR5NMm7PbWFRVbuBc5Ocv9KFnlBVh6vb5K6qfgYcoPcN7BbM1VgPeC3w3ar6wawLWazmAx34XeCHVfVfI9oL+EaSvd3WBLP27u5P0FuT/OqQ9lHbLMyDt9GbdQ0z63GeZNzmdmyTbAReBdw3pPnyJA8kuTvJxStb2Ujj3u+5HWt637UZNQGcx7F+zqreDz3JvwHnDWn6YFV9pbv/Jk4+O7+iqp5I8mLgm0m+U1X3TLvWE05WM71dKj9C7z+Gj9BbKnrb4FMM+dllPfd0knFO8kHgOHD7iKdZ0XEeYmpbWKy0JGcDXwJuqqqjA8376C0NHOs+c7mL3q6nszbu/Z7XsT4LeAPwN0Oa53Wsn7OqA72qXney9iRnAn8MvPokz/FE9++TSb5M70/zZQuacTWfkOQfga8OaVrxbRYmGOc/A64FXlvdYuOQ51jRcR5iVW5hkWQNvTC/varuHGzvD/iq2pXk00nWVtVMN5Oa4P2eu7HuXAXsq6ofDjbM61j3a33J5XXAd6pqYVhjkl9Jcs6J+/Q+4Ht4BesbrKd/DfGPRtQyyVYMKya9i5/8NfCGqnpmRJ95GOdVt4VF9/nPZ4EDVfWxEX3O6/qR5FJ6/00/tXJVDq1pkvd7rsa6z8i/6OdxrAet6hn6BF6wFpbk1+lddelq4NeAL3fv0ZnA56vqayte5S/8XZJX0vvT8xDwDnh+zTViK4YZ1QuwDXgRvT+rAXZX1TvnbZxHjVvmewuLK4A3Aw/lF6fdfgC4EJ6r+XrgxiTHgWeBG0b9lbSChr7fcz7WJPllehfyeUffsf6a53Gsn8ev/ktSI1pfcpGk04aBLkmNMNAlqREGuiQ1wkCXpEYY6JLUCANdkhrx/7SqgyI5a/ixAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(sel_df[\"normxpehh\"], bins=100)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1152a659-9ed2-49a0-bc5f-07301615c596",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "35812528it [20:28, 29153.70it/s]\n"
     ]
    }
   ],
   "source": [
    "pos_df = pd.read_csv(f\"{proj_dir}/results/fig_plotting_info/plotting_info.csv\", sep=\",\") #generated in prep_figs_info.ipynb\n",
    "\n",
    "offsets=dict(zip(pos_df[\"contig\"], pos_df[\"x_start\"].astype(int)))\n",
    "colors=dict(zip(pos_df[\"contig\"], pos_df[\"color\"].astype(str)))\n",
    "\n",
    "x=[]\n",
    "c=[]\n",
    "\n",
    "for index, row in tqdm(sel_df.iterrows()):\n",
    "    snv_id=row[\"id\"]\n",
    "    chrom = row[\"chr\"]\n",
    "    \n",
    "    offset=offsets[chrom]\n",
    "    color = colors[chrom]\n",
    "    \n",
    "    x.append(row[\"pos\"] + offset)\n",
    "    c.append(color)\n",
    "    \n",
    "sel_df[\"genomic_position\"] = x\n",
    "sel_df[\"color\"] = c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "480363e2-a129-4e6c-8dbe-7ddc50299b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_df.to_csv(\"sel_df_norm.csv\", sep=\",\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1179d8b2-bec1-45fa-a2b9-227c614122cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_380452/511402116.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bed[\"stop\"] = bed[\"start\"] + 1\n"
     ]
    }
   ],
   "source": [
    "bed = sel_df[[\"chr\", \"pos\", \"pos\", \"normxpehh\"]]\n",
    "bed.columns=[\"chr\", \"start\", \"stop\", \"normxpehh\"]\n",
    "bed[\"stop\"] = bed[\"start\"] + 1\n",
    "bed.to_csv(\"sel_df.bed\", sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "890085f5-f6fb-4ec7-896e-8cadc56c293d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "z_scores = stats.zscore(sel_df[\"xpehh\"])\n",
    "#sel_df[\"norm_z_scores\"] = z_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "6525bb75-c59f-4dfb-8e44-a02c881037cb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>gpos</th>\n",
       "      <th>p1</th>\n",
       "      <th>ihh1</th>\n",
       "      <th>p2</th>\n",
       "      <th>ihh2</th>\n",
       "      <th>xpehh</th>\n",
       "      <th>normxpehh</th>\n",
       "      <th>crit</th>\n",
       "      <th>norm_p_value</th>\n",
       "      <th>p_value_bonferroni</th>\n",
       "      <th>fdr</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>16981</td>\n",
       "      <td>NC_067195.1:16981</td>\n",
       "      <td>16981.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>526.758</td>\n",
       "      <td>0.042373</td>\n",
       "      <td>1417.040</td>\n",
       "      <td>-0.429771</td>\n",
       "      <td>-1.81653</td>\n",
       "      <td>0</td>\n",
       "      <td>0.069289</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>320430638</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>16982</td>\n",
       "      <td>NC_067195.1:16982</td>\n",
       "      <td>16982.0</td>\n",
       "      <td>0.012195</td>\n",
       "      <td>514.763</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1420.990</td>\n",
       "      <td>-0.440983</td>\n",
       "      <td>-1.87203</td>\n",
       "      <td>0</td>\n",
       "      <td>0.061202</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.973371</td>\n",
       "      <td>320430639</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>16993</td>\n",
       "      <td>NC_067195.1:16993</td>\n",
       "      <td>16993.0</td>\n",
       "      <td>0.036585</td>\n",
       "      <td>484.027</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1420.990</td>\n",
       "      <td>-0.467720</td>\n",
       "      <td>-2.00436</td>\n",
       "      <td>1</td>\n",
       "      <td>0.045032</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.888875</td>\n",
       "      <td>320430650</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>16994</td>\n",
       "      <td>NC_067195.1:16994</td>\n",
       "      <td>16994.0</td>\n",
       "      <td>0.128049</td>\n",
       "      <td>471.285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1420.990</td>\n",
       "      <td>-0.479306</td>\n",
       "      <td>-2.06171</td>\n",
       "      <td>1</td>\n",
       "      <td>0.039235</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.849029</td>\n",
       "      <td>320430651</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>16995</td>\n",
       "      <td>NC_067195.1:16995</td>\n",
       "      <td>16995.0</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>474.874</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1420.990</td>\n",
       "      <td>-0.476012</td>\n",
       "      <td>-2.04540</td>\n",
       "      <td>1</td>\n",
       "      <td>0.040815</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.859831</td>\n",
       "      <td>320430652</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117383</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477723</td>\n",
       "      <td>NC_067202.1:19477723</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>398.509</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.169044</td>\n",
       "      <td>1.53885</td>\n",
       "      <td>0</td>\n",
       "      <td>0.123841</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>317910191</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117384</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477730</td>\n",
       "      <td>NC_067202.1:19477730</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.036585</td>\n",
       "      <td>396.811</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.167189</td>\n",
       "      <td>1.53106</td>\n",
       "      <td>0</td>\n",
       "      <td>0.125755</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>317910198</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117385</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477740</td>\n",
       "      <td>NC_067202.1:19477740</td>\n",
       "      <td>19477700.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>409.008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>270.019</td>\n",
       "      <td>0.180337</td>\n",
       "      <td>1.58626</td>\n",
       "      <td>0</td>\n",
       "      <td>0.112680</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>317910208</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117386</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477770</td>\n",
       "      <td>NC_067202.1:19477770</td>\n",
       "      <td>19477800.0</td>\n",
       "      <td>0.390244</td>\n",
       "      <td>333.042</td>\n",
       "      <td>0.245763</td>\n",
       "      <td>207.029</td>\n",
       "      <td>0.206467</td>\n",
       "      <td>1.69596</td>\n",
       "      <td>0</td>\n",
       "      <td>0.089893</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>317910238</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2117387</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>19477785</td>\n",
       "      <td>NC_067202.1:19477785</td>\n",
       "      <td>19477800.0</td>\n",
       "      <td>0.012195</td>\n",
       "      <td>379.232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>236.080</td>\n",
       "      <td>0.205845</td>\n",
       "      <td>1.69335</td>\n",
       "      <td>0</td>\n",
       "      <td>0.090389</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>317910253</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35812528 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chr       pos                    id        gpos        p1  \\\n",
       "0        NC_067195.1     16981     NC_067195.1:16981     16981.0  0.000000   \n",
       "1        NC_067195.1     16982     NC_067195.1:16982     16982.0  0.012195   \n",
       "2        NC_067195.1     16993     NC_067195.1:16993     16993.0  0.036585   \n",
       "3        NC_067195.1     16994     NC_067195.1:16994     16994.0  0.128049   \n",
       "4        NC_067195.1     16995     NC_067195.1:16995     16995.0  0.006098   \n",
       "...              ...       ...                   ...         ...       ...   \n",
       "2117383  NC_067202.1  19477723  NC_067202.1:19477723  19477700.0  0.000000   \n",
       "2117384  NC_067202.1  19477730  NC_067202.1:19477730  19477700.0  0.036585   \n",
       "2117385  NC_067202.1  19477740  NC_067202.1:19477740  19477700.0  0.000000   \n",
       "2117386  NC_067202.1  19477770  NC_067202.1:19477770  19477800.0  0.390244   \n",
       "2117387  NC_067202.1  19477785  NC_067202.1:19477785  19477800.0  0.012195   \n",
       "\n",
       "            ihh1        p2      ihh2     xpehh  normxpehh  crit  norm_p_value  \\\n",
       "0        526.758  0.042373  1417.040 -0.429771   -1.81653     0      0.069289   \n",
       "1        514.763  0.000000  1420.990 -0.440983   -1.87203     0      0.061202   \n",
       "2        484.027  0.000000  1420.990 -0.467720   -2.00436     1      0.045032   \n",
       "3        471.285  0.000000  1420.990 -0.479306   -2.06171     1      0.039235   \n",
       "4        474.874  0.000000  1420.990 -0.476012   -2.04540     1      0.040815   \n",
       "...          ...       ...       ...       ...        ...   ...           ...   \n",
       "2117383  398.509  0.000000   270.019  0.169044    1.53885     0      0.123841   \n",
       "2117384  396.811  0.000000   270.019  0.167189    1.53106     0      0.125755   \n",
       "2117385  409.008  0.000000   270.019  0.180337    1.58626     0      0.112680   \n",
       "2117386  333.042  0.245763   207.029  0.206467    1.69596     0      0.089893   \n",
       "2117387  379.232  0.000000   236.080  0.205845    1.69335     0      0.090389   \n",
       "\n",
       "         p_value_bonferroni       fdr  genomic_position    color  \n",
       "0                       1.0  0.999998         320430638  #7bade3  \n",
       "1                       1.0  0.973371         320430639  #7bade3  \n",
       "2                       1.0  0.888875         320430650  #7bade3  \n",
       "3                       1.0  0.849029         320430651  #7bade3  \n",
       "4                       1.0  0.859831         320430652  #7bade3  \n",
       "...                     ...       ...               ...      ...  \n",
       "2117383                 1.0  0.999998         317910191  #078d70  \n",
       "2117384                 1.0  0.999998         317910198  #078d70  \n",
       "2117385                 1.0  0.999998         317910208  #078d70  \n",
       "2117386                 1.0  0.999998         317910238  #078d70  \n",
       "2117387                 1.0  0.999998         317910253  #078d70  \n",
       "\n",
       "[35812528 rows x 16 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_values = stats.norm.sf(np.abs(sel_df[\"normxpehh\"])) * 2  # Two-tailed p-values\n",
    "sel_df[\"norm_p_value\"] = p_values\n",
    "\n",
    "sel_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "90098524-0b89-4755-b86c-f9cfdc9f50ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bonferroni correction\n",
    "_, p_values_bonferroni, _, _ = smm.multipletests(sel_df[\"norm_p_value\"], method='bonferroni')\n",
    "sel_df[\"p_value_bonferroni\"] = p_values_bonferroni\n",
    "\n",
    "# Benjamini-Hochberg correction (False Discovery Rate)\n",
    "_, fdr, _, _ = smm.multipletests(sel_df[\"norm_p_value\"], method='fdr_bh')\n",
    "sel_df[\"fdr\"] = fdr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "385f577c-429e-4db5-aafc-b745ebd4540d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sel_df.to_csv(\"sel_df_norm.csv\", sep=\",\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ba34ab56-be2d-407e-8912-cee4343e634e",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = 0.05\n",
    "sig_bonf = sel_df[sel_df[\"p_value_bonferroni\"] < alpha]\n",
    "sig_fdr = sel_df[sel_df[\"fdr\"] < alpha]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3e538712-843c-4e42-90cf-5bbdca1f783c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>gpos</th>\n",
       "      <th>p1</th>\n",
       "      <th>ihh1</th>\n",
       "      <th>p2</th>\n",
       "      <th>ihh2</th>\n",
       "      <th>xpehh</th>\n",
       "      <th>normxpehh</th>\n",
       "      <th>crit</th>\n",
       "      <th>norm_p_value</th>\n",
       "      <th>p_value_bonferroni</th>\n",
       "      <th>fdr</th>\n",
       "      <th>genomic_position</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>85440</td>\n",
       "      <td>NC_067195.1:85440</td>\n",
       "      <td>85440.0</td>\n",
       "      <td>0.067073</td>\n",
       "      <td>17669.20</td>\n",
       "      <td>0.203390</td>\n",
       "      <td>1093.0900</td>\n",
       "      <td>1.20856</td>\n",
       "      <td>6.29241</td>\n",
       "      <td>1</td>\n",
       "      <td>3.125746e-10</td>\n",
       "      <td>0.011194</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>320499097</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>85469</td>\n",
       "      <td>NC_067195.1:85469</td>\n",
       "      <td>85469.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17846.90</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1216.4200</td>\n",
       "      <td>1.16648</td>\n",
       "      <td>6.08413</td>\n",
       "      <td>1</td>\n",
       "      <td>1.171255e-09</td>\n",
       "      <td>0.041946</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>320499126</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>85471</td>\n",
       "      <td>NC_067195.1:85471</td>\n",
       "      <td>85471.0</td>\n",
       "      <td>0.006098</td>\n",
       "      <td>17844.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1216.4200</td>\n",
       "      <td>1.16641</td>\n",
       "      <td>6.08378</td>\n",
       "      <td>1</td>\n",
       "      <td>1.173816e-09</td>\n",
       "      <td>0.042037</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>320499128</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>85477</td>\n",
       "      <td>NC_067195.1:85477</td>\n",
       "      <td>85477.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17844.40</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1216.4200</td>\n",
       "      <td>1.16642</td>\n",
       "      <td>6.08383</td>\n",
       "      <td>1</td>\n",
       "      <td>1.173450e-09</td>\n",
       "      <td>0.042024</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>320499134</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>85480</td>\n",
       "      <td>NC_067195.1:85480</td>\n",
       "      <td>85480.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>17844.40</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1216.4200</td>\n",
       "      <td>1.16642</td>\n",
       "      <td>6.08383</td>\n",
       "      <td>1</td>\n",
       "      <td>1.173450e-09</td>\n",
       "      <td>0.042024</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>320499137</td>\n",
       "      <td>#7bade3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041856</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>18613621</td>\n",
       "      <td>NC_067202.1:18613621</td>\n",
       "      <td>18613600.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2354.28</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.0760</td>\n",
       "      <td>1.29240</td>\n",
       "      <td>6.25489</td>\n",
       "      <td>1</td>\n",
       "      <td>3.977962e-10</td>\n",
       "      <td>0.014246</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>317046089</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041857</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>18613622</td>\n",
       "      <td>NC_067202.1:18613622</td>\n",
       "      <td>18613600.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2354.28</td>\n",
       "      <td>0.144068</td>\n",
       "      <td>114.8890</td>\n",
       "      <td>1.31158</td>\n",
       "      <td>6.33541</td>\n",
       "      <td>1</td>\n",
       "      <td>2.367110e-10</td>\n",
       "      <td>0.008477</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>317046090</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041861</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>18613632</td>\n",
       "      <td>NC_067202.1:18613632</td>\n",
       "      <td>18613600.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2353.65</td>\n",
       "      <td>0.093220</td>\n",
       "      <td>122.1220</td>\n",
       "      <td>1.28495</td>\n",
       "      <td>6.22362</td>\n",
       "      <td>1</td>\n",
       "      <td>4.858129e-10</td>\n",
       "      <td>0.017398</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>317046100</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041862</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>18613640</td>\n",
       "      <td>NC_067202.1:18613640</td>\n",
       "      <td>18613600.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2353.65</td>\n",
       "      <td>0.084746</td>\n",
       "      <td>132.2120</td>\n",
       "      <td>1.25047</td>\n",
       "      <td>6.07886</td>\n",
       "      <td>1</td>\n",
       "      <td>1.210400e-09</td>\n",
       "      <td>0.043347</td>\n",
       "      <td>0.000046</td>\n",
       "      <td>317046108</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2055506</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>18770708</td>\n",
       "      <td>NC_067202.1:18770708</td>\n",
       "      <td>18770700.0</td>\n",
       "      <td>0.012195</td>\n",
       "      <td>1196.57</td>\n",
       "      <td>0.440678</td>\n",
       "      <td>60.3998</td>\n",
       "      <td>1.29690</td>\n",
       "      <td>6.27379</td>\n",
       "      <td>1</td>\n",
       "      <td>3.523637e-10</td>\n",
       "      <td>0.012619</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>317203176</td>\n",
       "      <td>#078d70</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>996 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chr       pos                    id        gpos        p1  \\\n",
       "581      NC_067195.1     85440     NC_067195.1:85440     85440.0  0.067073   \n",
       "582      NC_067195.1     85469     NC_067195.1:85469     85469.0  0.000000   \n",
       "583      NC_067195.1     85471     NC_067195.1:85471     85471.0  0.006098   \n",
       "584      NC_067195.1     85477     NC_067195.1:85477     85477.0  0.000000   \n",
       "585      NC_067195.1     85480     NC_067195.1:85480     85480.0  0.000000   \n",
       "...              ...       ...                   ...         ...       ...   \n",
       "2041856  NC_067202.1  18613621  NC_067202.1:18613621  18613600.0  0.000000   \n",
       "2041857  NC_067202.1  18613622  NC_067202.1:18613622  18613600.0  0.000000   \n",
       "2041861  NC_067202.1  18613632  NC_067202.1:18613632  18613600.0  0.000000   \n",
       "2041862  NC_067202.1  18613640  NC_067202.1:18613640  18613600.0  0.000000   \n",
       "2055506  NC_067202.1  18770708  NC_067202.1:18770708  18770700.0  0.012195   \n",
       "\n",
       "             ihh1        p2       ihh2    xpehh  normxpehh  crit  \\\n",
       "581      17669.20  0.203390  1093.0900  1.20856    6.29241     1   \n",
       "582      17846.90  0.000000  1216.4200  1.16648    6.08413     1   \n",
       "583      17844.00  0.000000  1216.4200  1.16641    6.08378     1   \n",
       "584      17844.40  0.000000  1216.4200  1.16642    6.08383     1   \n",
       "585      17844.40  0.000000  1216.4200  1.16642    6.08383     1   \n",
       "...           ...       ...        ...      ...        ...   ...   \n",
       "2041856   2354.28  0.000000   120.0760  1.29240    6.25489     1   \n",
       "2041857   2354.28  0.144068   114.8890  1.31158    6.33541     1   \n",
       "2041861   2353.65  0.093220   122.1220  1.28495    6.22362     1   \n",
       "2041862   2353.65  0.084746   132.2120  1.25047    6.07886     1   \n",
       "2055506   1196.57  0.440678    60.3998  1.29690    6.27379     1   \n",
       "\n",
       "         norm_p_value  p_value_bonferroni       fdr  genomic_position    color  \n",
       "581      3.125746e-10            0.011194  0.000022         320499097  #7bade3  \n",
       "582      1.171255e-09            0.041946  0.000045         320499126  #7bade3  \n",
       "583      1.173816e-09            0.042037  0.000045         320499128  #7bade3  \n",
       "584      1.173450e-09            0.042024  0.000045         320499134  #7bade3  \n",
       "585      1.173450e-09            0.042024  0.000045         320499137  #7bade3  \n",
       "...               ...                 ...       ...               ...      ...  \n",
       "2041856  3.977962e-10            0.014246  0.000024         317046089  #078d70  \n",
       "2041857  2.367110e-10            0.008477  0.000019         317046090  #078d70  \n",
       "2041861  4.858129e-10            0.017398  0.000026         317046100  #078d70  \n",
       "2041862  1.210400e-09            0.043347  0.000046         317046108  #078d70  \n",
       "2055506  3.523637e-10            0.012619  0.000023         317203176  #078d70  \n",
       "\n",
       "[996 rows x 16 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_bonf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e3310778-4b32-4a9d-980e-8b3cb4668b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_bonf.to_csv(\"sel_df_norm_sig_bonf.csv\", header=True, index=False, sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9b3ae714-b500-49ca-94ad-aaca783eb88f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NC_067200.1    352\n",
       "NC_067195.1    286\n",
       "NC_067199.1    278\n",
       "NC_067196.1     38\n",
       "NC_067198.1     19\n",
       "NC_067197.1     13\n",
       "NC_067202.1     10\n",
       "Name: chr, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_bonf[\"chr\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "eaaad46a-ea8a-4ce1-a0a4-c4129c8b7225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.05713"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abs(sig_bonf[\"normxpehh\"]).min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "39ee0407-6b71-406f-959f-e3f683fd9660",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAb4AAABqCAYAAADDVTRkAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAAsTAAALEwEAmpwYAAAj30lEQVR4nO3deVxU9d4H8M/MAAMjm8AICooZiuTGo6NJLqCoLaaWZm5RLk+Eeq201TRu1+TxenOtWxmlWXrLNHGtLDfUbtiFCk0zr2SgoCgICMg+M88fwxnmzJyZc2Y9A3zfrxevH3PmLD+GM+f7W8+RaLVaLQghhJB2Qip2BgghhBBXosBHCCGkXaHARwghpF2hwEcIIaRdocBHCCGkXaHARwghpF0RNfA1NTVh+vTpGDVqFF5++WUxs0IIIaSd8BDz4Hv27MGAAQOwdOlSLFq0CGfOnMGAAQPMrh8SEoLu3bu7LoOEEEJapfz8fJSWlnK+J2rgu3z5Mvr37w8AiI2NRVZWlsXA1717d+Tk5Lgqe4QQQloplUpl9j1Rmzqjo6Nx4sQJAMDx48dRXl4uZnYIIYS0A6IGvgkTJqC2thaJiYmQy+UIDQ01WSc9PR0qlQoqlQolJSUi5JKQtq/40AkU3jsaxYdOiJ0VQpxO4i736kxOTkZqaioiIiLMrqNSqaipkxAnuHHPIIRe+Bk3YgYi9LefxM4OIXazFC9E7eMrKirCrFmzIJVK8eSTT1oMeoQQ55HVVLNSQtoyUQNfeHg4MjMzxcwCIQSAT1kJKyWkLaMJ7IQQKKrKWSkhbRkFPkKI3WhwDGlNRG3qJIS0DX6PP4oOVeW483guUFkmdnYIsYhqfIQQu1FTKbdz+zPw0//0wbn9GWJnhRigwEesRs1ahAgjWbIEg3J/g2TJErGzQgxQ4CNWk89OQsR/jkM+O0nsrLQaVPJvkTt3KW7LFMidu1TsrLiA1ig178TMx9Agk+LEzMecmyUifuD79NNPkZiYiISEBBQVFTn9eFRbsZ//jauslPCTLFncXPJfLHZWnO6PLRn4vUtf/LGFO8hHbV2PAE0toraud3HOXM+3/Da0zSmfYTt2w0ujxbAdu52fMTu1+oKcVkSFhYXauXPnCl5/0KBBdh+zJLKXVgvoUmITDaDVArpUq9Xmbd6tvdC5jzZv826r9vPLnFe1FVIf7S9zXnVGNllszaOjVHrKtFpAl7oh4/+pPduXS320WiblUA0PrRbQpW3IrU4RWg2gvdUpQvvfzkqthvk8BX6uDc3rNoh7WRZE3ZxXtRvn1VK8ELXG9+2330KtViMxMRGLFi2CWq12+jEDiv5kpcR+HgtS0Pv6eXgsSLFqu7s+2YgATS3u+mSjk3Km88eWDHSbNwW9r5+Hct4Mpx7LHN9GNSttywI0tazUmAJNrLSt6HizEJLmNOp6CSRWbi8zSt2ZxCg19k7KLJR28ME7KbNclSWriBr4bty4gYaGBhw9ehQKhQL79u0zWcfRN6mWNgdXqVqNXf+5Zff+CBDUcJuVCrUlaBQqJD7YEjTKGdnSu/38K/Bs/t0PDU49VmvVJPNkpcR6muYwoLEy5J2+lo8Juzdhx6De0ADYMSjGCblzrRnbMhBSU4cZ29yzKVTUwBcQEID4+HgAwOjRo3HhwgWTdZKTk5GTk4OcnBwolUq7j/n9pGfQJPPA95OewZlC7hIpsUwtkULbnALAJv/RqJD4YJP/aKxc9IXg/ez16o9JnZ/DXq/+TsqpzkfyOFTCE1oA5z3C9MvnfL0dPusWY87X2516fAAol+uOXy53XGC5cqseW78vwZVb9Xbv69jClbjj1xHHFq60aXs1dMM31AAuy4KgbU658L3fWh1IWYU7fh1xIGUVznfqCC2AOqlEd9516mh2u9lfb8PhgouYO/MBKNY+j7kz73dZnm3F/F11Uu4gn/rQfShVeCP1oftcmzGBRJ3Aft999+HDDz8EAOTm5uKuu+5y+jEPJ72Gw0mvWVwn+3I1vjt/G+P6BGBwD1+n56m1eePLAv3vKwEc7BCLgx1idQsycrH8nWmi5Muc817hmNDlBf1rZljT57//pE8/fugJp+ahy/8t1P9e56B9fniiBFoAf9wswZuT7bvBe+WseVgVPxMDInxs2v6N3S0DnU4telf/O9cQsrmhyRbfb61yxs5Czlhd0970nsUm72dey8fQLt1NlucLmPC/+WwWUr8/iBXDH8a8/nF259VeHd96Tv871/m8Ja4/tsTpCrTvuShP1hC1xhcbGwsfHx8kJCQgOzsbjz3mHsN49+VWoLZRi325FWJnxaK3vrmG5RmFeOuba2JnhYhA+EB5fkzrhy2tIBu+u+6AHLR9E3ZvsnnbpSf341ZdDZae3O/AHLVfok9nWLNmDTIzM/Hll1/Cy8tL7Oy0KrdrNazUXezf/qPYWbCa9zrbphlsPpuF8PeWYfPZLAfnqPUorTY/YGf/9h8xoe+bNp0T9mxrzrKTB+C3/gUsO3nAYfsUqqqR3STN9O1JBfQJ1jQ2sFJDjj4H7ekCYD5fdyd64HMHjuwrcYatp25ieUYhtp666ZLjZV+uRtqBImRf5n82G1ee1r6y16H5OZdTgBdnbcG5nAL+lTm2daZnj+7CrboaPHt0l1OP01qtfXUvKstrsPbVvdZv+0rztg48n9bmHEOjVoO1Occcts839hZieUah1dvN/24HDhdchIajzj5h9yacvpYPQBcgm7S6wi2TGlp4ZCdu1dVg4ZGd+m291y3mrGF235QK73WL0X1TKmu5YfD8/PefoIWuC4DJg1DM5ysEE/itPYYjUOADkH6iBHk365F+wj2fRZZX0sBKne2bX3VNvd/8WsG7rrk8xYcvxYsztzgkP68//S9kZ17C60//y4ZtnTtwRa3VslJHcMdapGHh8Mqterx//AY2Hb9hUli8/u/z7A0N2mNtLbxYonznVXivWwzlO686dL/WaLKxweVC2Q2z7x0uuIinvtoGn3WLkbDDuuk+hwsuslJDxTVVrBTQBSDj4Ml4KXMP7/FGfGbbjQjSsg7hcMFFpGUd0ufDVYGQns7QTly5VY9jFyoxOsYf3YLlFtdlvshcX2ghtUD9uicuWZNFs8puVrFS67ZtfU8UZy4+C4/sdIuBDACwK7sM5TVq3KouQ7CvB4rKGwEAX5+tYK2Xt9P8cJXszEuovu2YoT3MYA+m+dC4GVGo7ptSUVxThTCFH/JTVjgkb45SUMU96MWWgS7LTh7A2z9lcr7HBB4u1Q38n2t28RWL728+myUonxN2b0JVYz1OX8tHyaK/865vD6rxGRHa3MnXPJp9uRor9hVi4+Fis+ts+O46lmcUWj04gDk2H8Mmyz0/lyPvZj32/Mx/93yNlp0asmbAj1Rm7RReHb6+Hb6mT3uaRhlCS5+Dw7qxUnPCFH6s1JW1Okc05ZfXqPVp54CWKRm3qo0moXOdNIb7KRVWEDH3v2Oa8ZgaiiVC+vOYmk+C33gszyjE+8d0ozF3/ecWlmcUuuVcX3O1M0u4miCZc++RngPMbldeV4MRn63HiM822FwTS/3+IOdyw5pp902pdhdirMEb+DQaDe9Pa2YcQD40aO70tlAf/vpsBfJu1puUeBn7civQoAZKqpqw+SR3kGIGBVgaHAAAHtKW9MqtenzY3DTLx3B0akmV7gLFpK6gUbdcBF+cuUVw8+fbqQdQWV6Dt1O5L1j/eCkD2ZmX8I+X2JNjmYD55l++QHbmJfzzDe4vHADeL/FTX23TNzcxuDr9mRIxX8m4sr6OlS46ousbXHRklz4/lgItE7xW7tf1J606KPy+toZN+d9yNF9bCohchaxTl1qCV22jdU28N4rYBS9zhZQ3/8I9H5Sr+c4c4/48S4Muwr27AACKKnTfD6GjXM21gHSWd7a4nbOa84z79ZgCFpeFR3bCe91iLD1peuMQRnFNFbKLryC7uMBizdCSgaFdedcxbHp1Bd6mTg8PD0gk3CV3rVYLiUTikluNOYtxADH8Gvv5eGDDd9dRWq1GiK8Mz49rOZmr6zWs1BK10bWBaXY0tPX7ErPNkIZNj8cuVNo9fJ3piJ8UG+iyeYpMs6eQ5s+mBjUrNfRwnxWoqtBdjAr+exPx4UsRpPTFntxl+kEQleW6msClc+Zr0i9l7jVZ5r1uMeRSGW4/v0bfzGTY3GQ47++Z2OF4KXOPvp+mqLrC4t9Uo27Up8tOHtAPaGDS+d/twIWyGyis4t7PZ6dvsc61Ow3CzgLjoPbvvGrc3y+QteyLbHatxpaBGkIZj3tYMuMj1Nc04ex//sR3l97ULy++6tjn+oW/t0zwoIu3vrkGCVquBal7CjEsyhfXKxqQV9KAKKUX+oQr8N3526gzE/gfD52KjVfeNll+unkun61BhI9xwUBIrbBSQHMmAIT4+OL0tXzOvJ82M0eRK08MhcwTNepG1mftKryB788/W+5pqdVqMWDAAJw9e9ZhGTh37hySk5Mhk8kQFRWFLVu2mA20rmZYOzKulVU0N/0wqTWOXag0Cbh5N+uRd7MEI3r64v5+gfj21wr8O68aw6LYgamwjPsktaYPj7EvtwL5pfWYOiSYtR9DjprM7ymXobFeDU85/50IfXzlqKmqh0QqMakJMEHPUFkJd6m7ycx9MTusWwK1ma9avUZtUhqP/nAFCo0C28SMD1DZ0NJfNb5HX879cVnHMaKQCaBcAx6yL1cLKmABYJ039/cLxL9Ol7Le7xFiOmXodo14rTb1NU2s1Fn4mkUN3a7VIFAhRUXz56LRsmu5eSUNvAPNPCTcl9aHvnwfZc+uFpwXPq646xDj899/QmltNWcgS9ixES+oRpvd1nvdYoyNjEaIjy92/P4TpvcepC8MujroAQICX2RkJOu1TCYzWWaP6Oho/PDDDwCAOXPmICcnB4MHD3bY/h3JMLgY2nrqJmaP6KR/ba75g6k9dvCSwMdTwtlMdOpSNYI6eOi/aIZfOACoM3N9YEak5t3UpV4C73R7prAWU4e0vDYu/TN9evtyK7Dfjgn9jfVqfbpy0Rc4nJHLep+puVVV1urX1ai1eOmJjwXt35q5XuaCHsO4RFvA8VRxw6AHADsM7v7CDCR4dlAC0kZOAKDrA2QGARgf3bgJLkzhh+KaKgTJFZiwexMGSMbD/O2Add4/VqxvpgN05839/QJxp559tIKyBqw6WIQ7DVpIAIR3dL97c3L1753LKUBflbDrjqWBHEJVOKkwUNOkC5jWNNlawrRCuMp/y8xPqeKbImL4N7s638ZEH9zi6dnyxZPL5ejalb89WCxMX8lXxiPZDEp/3/5aYXYACFNrvNOgtdg34og7xnC0EpplOPCh0sIX3paSGVdAMg56jLKSan3QY9RUCWuGecdMf6CxG/34axa2XJS0aOm3Mexb8l63GN7rFlsc+WbcBHezRlfYKauvweGCi+CaKWEcBg2DHoOr765R3dJMqgVQ2Dw60x3Ehy/F/u0/YuEjpvPPjPtzLVmfc1xQs6ZYUyB8bLxZgjvgKgTaSsx2PbeYzrB//3689tpr6NWrF4KDg1nvpaenIz09HQAc8nQGR6jkaN5kaoPmBp2kOrHfxF5MM6uHFOgS6MF5EbXV2lf24loB/70I7dVQLyzS549xXpNawo6N8JTYX5bU8hQxfDwlkACo4RlYYlxAaw3MTVa/+keJ4EAlk0qg4RldCrhm9CAXMZr23JGlz0HoFAhb8Qa+1FT2DP/a2lqTZStW2Df/ZeLEiZg4cSIWLVqEgwcP4tFHH9W/l5ycjORk3U1tVSqVXcdxlCqO/ha+ye+tYexrk4a75mBJdtpnvOt8/t5JW7PkeE4uZgodQGEJ34VR6EjKIjeqzdlLo9byBipbbztH3M+zR3eJG/iuXr3Kej19+nSTZfaor6+HXK4bjOHv7w8fH9vuDk/EUVfs2NF3hHD5dbqjnmlBWgNH3gmJC2/g+/hjYYMLbHXo0CGsW7cOANCzZ0+MGzfOqccjhLQ+NV3EzgFpS3gDn5AJ6lKp7f0akyZNwqRJk2zenhBCCLFGu5/ATgghpH0RfQI7IYQQ4kqiT2AnhBBCXEn0CeyEEEKIK1HgI4QQ0q64xQR2QgghxFVEn8C+ePFi5OTkYODAgdi4caPD9ksIIYRwEXUC+88//4w7d+7g1KlTmD9/PrKzs932yQyEEELaBt4+vl27drFeX7zIvnP9hg0bbD54VlYWxowZAwAYM2YMTp8+bfO+CCGEECF4A9+8efNYr+Pi2DcONe7vs0ZFRQX8/XXPtgsICEB5uel9H9PT06FSqaBSqdzm6QyEEEJaL97ApzW6WSjfa2sEBgaisrISAFBZWYnAwECTdZKTk5GTk4OcnBwolUqbj0UIIYQAAgKf8e3K+F5bIy4uDkePHgUAHDlyBEOHDrV5X4QQQogQgubxabVaaDQa/T05jV/bauDAgfD29saIESMglUoxZMgQu/ZHCCGE8OEd1VldXQ0Pj5bVtFqt/jVzk2p70BQGQgghrmTVTaoJIYSQ1s7qm1RrtVqUlpYiJCTE7toeIYQQ4mqC79VZUVGBpKQkeHt7IzQ0FD4+PkhKSkJZWZkz80dIu0fly9anWl0tdhaIBYID35w5c1BbW4vc3FxUV1fjl19+QX19PebOnevM/BFCQJGvtdlfckDsLBALeJs6GcePH8f169fh4+MDAIiJicHWrVvRpUsXp2WOEEJao+v118XOArFAcI0vOjoa+fn5rGVXrlxBdHS0zQf/5JNPMHLkSAwZMgTvvfeezfshhBB72HMjDtL6CK7xJSYmYty4cUhKSkLXrl1x9epVbN++HUlJSdiyZYt+PWuaPmfOnImnnnoKarUaAwcOxIIFC6zLPSHWsqHV0EsqQ4PGvjmr5rJCl1v3cLjsCO4PGSd2NoiLCK7xZWVlISoqCllZWdi5cyeysrJw991344cffsC2bduwbds2bN++3aqDe3p6AgAaGhoQExNjXc7dmDs93Xfl5AjB64YHeiC8oyciOno6MUety9jIaKwfPcUp+xYa9JLj6VZ9zvZr9a8my+h70HYJvkb/7W9/w/Hjx01+kpOT9b8fO3bM6gysWLECPXv2xKBBg6ze1tU6KmTo4KWrMgT4mP/o/jdeiahOckyKDURUJ7mgfSfHK6H0kzkkn5Z4NGfbi+NQ80eHYf6oUKSMCnXY8QbH90RY144O2VdkLyX8Oyp413PU8QDgwJQUzOsfx7+iAamAauW7Yx7H2MhoDA7rxrtut2Bh51CIr3Xnz6TYQKvWb8veHfO4yTJHfg+IexEc+KZMmYJXXnkFjY2NAHTTG6ZNm4a//vWvvNsWFxcjISGB9TN9+nQAuqc7/PHHH9i1axdu3bplsq27PJ1h5eQIvPBAZyx9OBwrJ0fgpQfND+rpFizH7OFKDO7hi9nDlbwXpOR4JboFy/HowCCH5jlK6QUA8Gz+L3tKgTceicDKyRFInSS8JggAIX4yzmDJSwK8/s9pGJzQ0+StE0WrWK9nLBgJqUyib47s4Me+4Cs6eKOyvIb3kMzxXlj9iA0ZbiGzEMAi/biDa5jCD++MmcpaFuytQHd/9v92Xv84HJiSglMzF9uVR0Ol1cKaY5V+MqycHIHBPXwddmync/LAVuPCzYiereizcSBPqTu1VzmP4L/yzJkzyM3NxeDBg7F582b069cPgYGB+OWXX3i3DQsLQ2ZmJutnx44dqK+vBwB4eXlBoVBALjct2brD0xkCFcJPBq7g8Py4zha3YUr0Qkv2QvMxe0QnAMCcEboa6JwR7M8v3EJTDvN3eEiBqE5yTB4YZHWwBIDZixPRVxWJNf/i7vsNiwjUpynLHsTxK/+HF/7+CPw7KpCy/EHMWDASMg8pZiwYib+8MV4XQHkugszxJj5xL2t5TKx1+X/bKIAxxkZG45PxT2JspOnArvyUFZjXPw6Z05/D2MhoZE5/DkUL0rD1oSRBx5RBgjCFn1X5tJbcg/98lrWRGRRCat9c7u8XKGg9RwZI48KRJc44R+qWrEfV82tRt2Q9ZvR2/xY4ewi+onfp0gV79+6FRqNBcnIyHnzwQXzwwQfo0KGDzQdftWoVEhISMGzYMEybNg2+vu5Zyorv5S94XVuCgz0GRPiYLEuOV7LywdRAjQPr+P6BZvfbyV8XFMMCPDm3ZUi9LfeD9FVFWnz/ix9fwYmiVfjix1f0yyY+cS8OnHsdE5+4FynLHsSxgjSkLHuwJYDaOCJk01cLzb7nLWOP85rRe5DZJs4DU1IwtEt3LIt7wOz+hnbprl+PeS3EwLCuyE9ZoQ+qXMHVkNxDd2G31KTO1PhlEt16Dxn83831H/7t0Qgo/XSfSaBCJrjJ3t0Y1765cDVz8gloLgyfLaqBr5x9GbWmCTnSTxfsYoJCMaVXrMV1JQZpcU0V777HRkbj3TGPI9hbYdJCEezN3WVw+lo+JuzehGdih7OWS8F/LrYmggNfbm4uVCoVevTogX379uHYsWOYMWMGKioqbD74G2+8gczMTPzwww+YP3++zftxBMNCsPGJe/5aLec24YGCB8WyOPIicqdBo//dU9bSbCqEpfUe6h9ocpHk0m/BRPh1NA2+ziSVtpTi392XYtW2Chl3oF47ajKCvRV4d8zjqFuyHh8/9ATvvoQGMz7MBSXYW4G3Eh4FACyLewBjI6P1wdVcM/NTw0IQ1UmO0THmC2fM5+Uhk5gUYiydA48O7IioTnI8PjgIs4e3vgE23jIPzsKLcaCztg8XACprNPp05tBg/aCw5HilVU3In4xPwtjIaLw/bjq2njttcV1Vc3+wKqyb2fMY0AXGGb0H4cCUFOy9dAa36mrQK6gTa53djzzNGcjSsg7hcMFFpGUdYi2vWbIeP17PF/ZHOUBMUJhT9y848CUmJmLJkiXYu3cvHn74YZw5cwYKhQL9+vVzZv5cZm5zcyDXiWvuojJ/dJjgUZNMkAwP9MDs4Up9/xuTMpj+QON+QaZU6SuXskp+o2P89V+6ORZqZtYyV0s05n9XGA6eS7V6/2Mnx9qYM6BrlO4iHNmrE2+N0tjXU7mnzMzrH4eiBWk2XQTtdWBKCuqWrEfRgjRWDdGwxthgpvtOyP/pgb4B8PGU4IG+AVblS+g54K7WjpoMgF3YCVP4OeR/PLynL6QSXdotWK4fFMb1WVkKUob/5xXDH7Z4zItlN/RpHyX3GIPu/kGoNSi4GRagDJvfmeMaMy5wGQr3DbSYP0fy9XLuOSe4ypKdnY0ePXroX3fo0AGbN2/G/v37nZIxV2O+5Obes9e1iiZWyvS/GTPsD0zNKIQGutLJzKHBOHahEqNj/PHV2QoUlTeiS0dP/ZfOHRn3qUX2VKLgUgkieyqx/J1pNu/35bcmY+v6o5i9ONHqbR1VS+PiJXXeqFwPKdCkYS8zbub2lAKNGl06Z4RSf750C5ZbVQsx1/wplQCaVjLxMEiu0Ae4r6cuQFrWISyLe8Ah/38Pqa4P0FI/oBTQf3e/nroAL2XuQXVDAy6UFZvdZl7/OCw8spO1bGxkNA4XXMTYyGj8WnodlQ31UHjK8VbCI3jqq09xpaoccpkHFv7PSJwtKTIJWMYBzjjYdfDwwp2mBnTw8OJc39D746YjLesQ+ivDsTbH+hH8hiL9OqKgqpzzPXOB15EEBz7DoGdo4sSJDsuMWLj6+n3lElTXa+Erd0wvv9YoFUJjkBoG5vH9A/UXNVdLjlci/QT/6FrjEZsA8GnmEofkwdJgGRPMLHGDf2Pm9Of0TTqOdHdgiEP3Z2hucyDLu1mvXzZ1SDBrnUZNS2qpIMfHXEEvyFeG0irHT+S3hadEikatxmT54LBIvJXwCCvAcV3MY4JCcaHsBmKCrC80GhdAuBh+d9OyDiG7+ArGRkazAl+QnH9qjmG+O/3zVQBATWM9hnbpjotPt7S0nL6Wj7MlRYLyb+irx+brCwV8DD/HHoEhWHpyH8J9A3GhuSZqjYtPp8J7nemI5iC5wmzgdaT2MXbVAh9PCeaOML1AzByq6zuZOZT/YrZycoT+xxy5jJ0KMaK5OcV45Jgjm6DMHcMcd2z2kivMNyXNmN88KnT+SP0yS6Vae7w/brrF919QjYanRIoXVKOt3jfzP/dsPn88nT/l08TkgUGI6iS3uW+bi39HhW7aiZXlS3+5t34OZExQqL4J79TM5wXV6t4fN13ft2YJU/A1HOVq3D3Bx1zz4bWFaVbtZ9XISQj2VmDVyEkm75nrm+Nj3KQu1N5LZ1DZUI8Iv0Cz63ANoJFJJGYHE9UtWW/1Z2Irx53BrVTSfSGcF3N7SsxcnhqutLqWxtec4giuOIazrft8Hp5//EM01qsR1jWQ9V7KsgeRsuxBu4/B1BKNL16GFUq+i0fayAlIGznBrnw81C8Q352/jXF9rOuvs2RET1+cusT/GB3D78TyjEL98vCOnigqb7Q4PcacA+deBwBcKyjDzvTv8XjycHz+3kne7VYMf9iuvjqhhZ+ZQ0Nw7EIl+nTxwflrtfqmYz6G54ujClrMQJW9l86Y/O3MeensJkKu45lrOSlakMaq1THB35ndDUK1u8A3KTYQ+3Ir9K9dVYNxdCAlLfqqIrFh59NW9/sNDuuG7OIrgu6eYu7iVbtkvVV5tdfgHr4On3h+f79AQYHPUHI8uyDH/L7DxjwYFlCEBD5XDUIy/N5a87k7o1WBCTBcgcYZx2P64bhu1mDL8YzXf0E12u6+QluJ2tSp0Wjw4osvIjExEVOn8s+3cYTBPXxZoyKJfbimZkhFmP3M9PtZM8ozUO7DSolwhs3trhj9aTihuq1PrmY4+0YGfJibNHwy/kmrt61rLhAy82ON58kCuhYQZp6hLXMp7SFqje/LL79ETEwM1qxZ49LjDu/pi3/nVWNYlHtOmHd3UUov5JU0wMdToi/xv7svBf94aTdKr1ciZbn9TYuu4OrmIWcL8ZWhtFpt9T073c3YybE4nJHLWvbxQ08ImltpjwERPjhTWMt5UwhXkcs8UK/WjfwWMkndmRxRizw0daHFwTPz+seJMoVI1MB38OBBKJVKJCQkYNasWXj66addcty20K8lJq6pGH1Vkfj0uGNGbbqKswa5iIXv1niuEKT0RVmJdc2mxpa/Mw2PPDUUr87+BFXltRgcb3qfV2eYOiQYU4e45FBmrRs1WT+lQewanyO463dM1MB348YNDB8+HKtXr8aYMWMwceJEhIa655w0QtqySbGOGTSzJ3cZACA+fCmOXvs7ZADUAI5634Oxdb/hsPc9SKz7Tb8cMJ32AugKUrbcGMHdeUikaNJq4CHh7mWyVAOa0XsQdvz+E6a7WVPv+ZWb0aO8CnVSCbw1WvzYNRRwcd+3tVwS+IqLi/VPY2CEhYUhICAA8fHx8PDwQFxcHPLy8kwCX3p6OtLT0wFA1KczENKWOXrQTGRPJWTXdP3oMgDj6n6DpDkFWpa3N0emLRI8b86YK5p7bdGjvAoSAN4aLSQA7r1q/bw+V5NotVrR7sWwYcMGhIeHY+rUqRg/fjw++OADRESYnwunUqmQk5PjwhwSQmyllUhYT5nn+l0i3uWHOIjh/9md/q+W4oWoozrnzZuHzz//HMOGDcO9995rMegRQloXNXQXQTWYZs2WlLRNzP/b3Ynax+fn54eMjAwxs0AIcZK8zn3Q+/p55HXug7oHJuCuTzbiz6eew4CP/y521oiTtJYm7HZ/yzJCiHN4rlyB3zv3gefKFYjdsgoB6hrEblll031rCXEkCnyEEKe4e+5k9L52DnfPncxa/muvYdA0p6TtaQ1NnRT4CCEuNeDi95BqtRhw8Xuxs0Ic4FyPbqzaOzV1EkIIadN65V9ljdilGh8hhJA2zav56cTMHXqpxkcIIYS4GVED35UrVzB69GiMHDkSO3fuFDMrhBBCbFDs3wFAy1Pnyzq4/9NORA18q1evRlpaGo4fP46PPvoITU1NYmaHEEKIlW5t+xQ/xd6DrNdfxk+x9+D6Z9vFzhIvUSewX758Gf3794dMJkNoaCjy8vLQu3dvMbNECCHECn0nTgYmNk9ZWbFa3MwIJGqNLzo6GidOnEBNTQ1Onz6N8vJyk3XS09OhUqmgUqnoJtWEEELs5pKbVJt7OsPGjRvxzDPPoLGxEVKpFBs3bkSPHj3M7ickJATdu3d3cm7dQ0lJCZRKpdjZEE1r+/spv+6VB2ft2x0+NyJMfn4+SktLOd8T9ekMjNraWiQlJWHXrl2QSCT8G7QD7f1JFK3t76f8ulcenLVvd/jciP1E7eP76quvsGbNGshkMqxevZqCHiGEEKcTNfCNHz8e48ePFzMLhBBC2hmawO6mkpOTxc6CqFrb30/5da88OGvf7vC5Efu5RR8fIYQQ4ipU4yOEENKuUOAjhBDSrlDgI4QQ0q5Q4COEENKuUOAjhBDSrlDgI4QQ0q78PzfEYA+xo1huAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 453.543x113.386 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot the results\n",
    "#fig, ax = plt.subplots()\n",
    "\n",
    "# Create the Manhattan plot figure and axis\n",
    "cm = 1/2.54\n",
    "fig, ax = plt.subplots(figsize=(16*cm, 4*cm), facecolor=\"white\")\n",
    "\n",
    "# Define three shades of grey\n",
    "shades_of_grey = ['#808080', '#A0A0A0', '#C0C0C0']\n",
    "mlm_colors= [\"#078d70\", \"#7bade3\", \"#3e1a78\"]\n",
    "\n",
    "# Create the colormap\n",
    "custom_cmap = ListedColormap(mlm_colors)\n",
    "\n",
    "#plot position of the invadolysin locus\n",
    "#plt.axvline(x=invad_loc, ymin=-3, ymax=3)\n",
    "\n",
    "# Iterate through the groups of chromosomes and plot each group with a different color\n",
    "\n",
    "for i, (chrom, group) in enumerate(sel_df.groupby(\"chr\")):\n",
    "    color = custom_cmap(i % custom_cmap.N)  # Get a color from the colormap\n",
    "    ax.scatter(group[\"genomic_position\"], group[\"normxpehh\"], s=2, alpha=1, color=color, label=f'{chrom}', zorder=2)\n",
    "\n",
    "# # Plot the significant data points (Bonferroni corrected) in red\n",
    "ax.scatter(sig_bonf[\"genomic_position\"], sig_bonf[\"normxpehh\"], s=2, alpha=1, color='red', label='Sig.', zorder=3)\n",
    "\n",
    "# # Set the x-axis and y-axis labels\n",
    "#ax.set_xlabel(\"Genomic Position\", fontsize=12)\n",
    "ax.set_ylabel(\"xpEHH\", fontsize=12)\n",
    "ax.tick_params(axis='y', labelsize=8)\n",
    "\n",
    "x_ticks =list(pos_df[\"mid_tick\"].astype(int)[0:7])\n",
    "x_labels =list(pos_df[\"contig\"].astype(str)[0:7])\n",
    "y_ticks = list(range(-9, 10, 3))\n",
    "ax.set_xticks(x_ticks)\n",
    "ax.set_yticks(y_ticks)\n",
    "\n",
    "ax.set_xticklabels([])\n",
    "                     \n",
    "#ax.axhline(y=sig_bonf[\"xpehh\"][sig_bonf[\"xpehh\"]>0].min(), color='black', linestyle=\"dashed\", linewidth=0.5)\n",
    "#ax.axhline(y=sig_bonf[\"xpehh\"][sig_bonf[\"xpehh\"]<0].max(), color='black', linestyle=\"dashed\", linewidth=0.5)\n",
    "\n",
    "#plot invadolysin line\n",
    "#ax.axvline(x=invad_loc_x, color='tomato', linestyle=\"solid\", linewidth=1, zorder=1)\n",
    "\n",
    "\n",
    "# Add the legend outside the plot in the upper right corner\n",
    "#ax.legend(markerscale=3, fontsize=12, loc='upper left')\n",
    "\n",
    "# Adjust the layout to fit the plot in the figure\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "# Save the plot as an SVG file\n",
    "#fig.savefig('nw_vs_se_xpehh_norm.svg', format='svg')\n",
    "\n",
    "# Save the plot as a 300 dpi PNG file\n",
    "fig.savefig('nw_vs_se_xpehh_norm.png', format='png', dpi=300, bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b72e973-48c7-4460-9382-e6f3f5560339",
   "metadata": {},
   "source": [
    "# regions under selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "afcd8039-527b-4be5-b49f-edd81e9d21c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_380452/435702102.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bed[\"stop\"] = bed[\"start\"] + 1\n"
     ]
    }
   ],
   "source": [
    "bed = sig_bonf[[\"chr\", \"pos\", \"pos\", \"normxpehh\"]]\n",
    "bed.columns=[\"chr\", \"start\", \"stop\", \"normxpehh\"]\n",
    "bed[\"stop\"] = bed[\"start\"] + 1\n",
    "bed.to_csv(\"sel_df_norm_sig_bonf.bed\", sep=\"\\t\", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "d0ff320e-211b-4351-b992-f3fc68514f0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "!conda run -n bedtools bedtools cluster -d 1000000 -i sel_df_norm_sig_bonf.bed >sel_df_norm_sig_bonf_clusters.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "8b29a617-d469-497a-a1cd-e5ccf75bf3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_df = pd.read_csv(\"sel_df_norm_sig_bonf_clusters.bed\", header=None, sep=\"\\t\")\n",
    "r_df.columns=[\"chr\", \"start\", \"stop\", \"normxpehh\", \"cluster\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "dd7cf76d-8849-4606-9c5e-448f692482c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "nxpehh = list(r_df.groupby(\"cluster\")[\"normxpehh\"].median())\n",
    "start = list(r_df.groupby(\"cluster\")[\"start\"].min())\n",
    "stop = list(r_df.groupby(\"cluster\")[\"stop\"].max())\n",
    "counts = list(r_df.groupby(\"cluster\")[\"normxpehh\"].count())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "fe80b7e2-1710-4ffc-98b0-d49e06bc4264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>cluster</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>med_norm_xpehh</th>\n",
       "      <th>len</th>\n",
       "      <th>sig_snvs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>1</td>\n",
       "      <td>85440</td>\n",
       "      <td>86575</td>\n",
       "      <td>6.168670</td>\n",
       "      <td>1135</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2450744</td>\n",
       "      <td>2702312</td>\n",
       "      <td>6.369370</td>\n",
       "      <td>251568</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>3</td>\n",
       "      <td>15451924</td>\n",
       "      <td>15451973</td>\n",
       "      <td>6.432380</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>4</td>\n",
       "      <td>64303350</td>\n",
       "      <td>64481575</td>\n",
       "      <td>6.192720</td>\n",
       "      <td>178225</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>5</td>\n",
       "      <td>67218409</td>\n",
       "      <td>67220075</td>\n",
       "      <td>-6.708680</td>\n",
       "      <td>1666</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>6</td>\n",
       "      <td>69543706</td>\n",
       "      <td>70097099</td>\n",
       "      <td>-6.810760</td>\n",
       "      <td>553393</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>7</td>\n",
       "      <td>72947731</td>\n",
       "      <td>72947748</td>\n",
       "      <td>6.159410</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>8</td>\n",
       "      <td>335206</td>\n",
       "      <td>335218</td>\n",
       "      <td>6.191205</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>9</td>\n",
       "      <td>92138959</td>\n",
       "      <td>92139077</td>\n",
       "      <td>6.151370</td>\n",
       "      <td>118</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>10</td>\n",
       "      <td>1449094</td>\n",
       "      <td>1449134</td>\n",
       "      <td>6.114800</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>11</td>\n",
       "      <td>16497345</td>\n",
       "      <td>16497348</td>\n",
       "      <td>6.326160</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>12</td>\n",
       "      <td>46292383</td>\n",
       "      <td>47002311</td>\n",
       "      <td>6.099690</td>\n",
       "      <td>709928</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_067198.1</td>\n",
       "      <td>13</td>\n",
       "      <td>34371482</td>\n",
       "      <td>34372132</td>\n",
       "      <td>6.151620</td>\n",
       "      <td>650</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>14</td>\n",
       "      <td>28684454</td>\n",
       "      <td>28684894</td>\n",
       "      <td>6.791500</td>\n",
       "      <td>440</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>15</td>\n",
       "      <td>46492141</td>\n",
       "      <td>46551457</td>\n",
       "      <td>6.318750</td>\n",
       "      <td>59316</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>16</td>\n",
       "      <td>503560</td>\n",
       "      <td>703283</td>\n",
       "      <td>6.174620</td>\n",
       "      <td>199723</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>17</td>\n",
       "      <td>2751101</td>\n",
       "      <td>2751781</td>\n",
       "      <td>-6.243965</td>\n",
       "      <td>680</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>18</td>\n",
       "      <td>6627231</td>\n",
       "      <td>6627311</td>\n",
       "      <td>-6.259730</td>\n",
       "      <td>80</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>19</td>\n",
       "      <td>9584838</td>\n",
       "      <td>10269525</td>\n",
       "      <td>-6.231310</td>\n",
       "      <td>684687</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>20</td>\n",
       "      <td>18613602</td>\n",
       "      <td>18770709</td>\n",
       "      <td>6.239235</td>\n",
       "      <td>157107</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr  cluster     start      stop  med_norm_xpehh     len  sig_snvs\n",
       "0   NC_067195.1        1     85440     86575        6.168670    1135        14\n",
       "1   NC_067195.1        2   2450744   2702312        6.369370  251568        35\n",
       "2   NC_067195.1        3  15451924  15451973        6.432380      49         9\n",
       "3   NC_067195.1        4  64303350  64481575        6.192720  178225        35\n",
       "4   NC_067195.1        5  67218409  67220075       -6.708680    1666       129\n",
       "5   NC_067195.1        6  69543706  70097099       -6.810760  553393        62\n",
       "6   NC_067195.1        7  72947731  72947748        6.159410      17         2\n",
       "7   NC_067196.1        8    335206    335218        6.191205      12         6\n",
       "8   NC_067196.1        9  92138959  92139077        6.151370     118        32\n",
       "9   NC_067197.1       10   1449094   1449134        6.114800      40         3\n",
       "10  NC_067197.1       11  16497345  16497348        6.326160       3         2\n",
       "11  NC_067197.1       12  46292383  47002311        6.099690  709928         8\n",
       "12  NC_067198.1       13  34371482  34372132        6.151620     650        19\n",
       "13  NC_067199.1       14  28684454  28684894        6.791500     440        61\n",
       "14  NC_067199.1       15  46492141  46551457        6.318750   59316       217\n",
       "15  NC_067200.1       16    503560    703283        6.174620  199723        13\n",
       "16  NC_067200.1       17   2751101   2751781       -6.243965     680       114\n",
       "17  NC_067200.1       18   6627231   6627311       -6.259730      80        13\n",
       "18  NC_067200.1       19   9584838  10269525       -6.231310  684687       212\n",
       "19  NC_067202.1       20  18613602  18770709        6.239235  157107        10"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_df = r_df[[\"chr\", \"cluster\"]].drop_duplicates().reset_index(drop=True)\n",
    "cluster_df[\"start\"] = start\n",
    "cluster_df[\"stop\"] = stop\n",
    "cluster_df[\"med_norm_xpehh\"] = nxpehh\n",
    "cluster_df[\"len\"] = cluster_df[\"stop\"]- cluster_df[\"start\"]\n",
    "cluster_df[\"sig_snvs\"] = counts\n",
    "\n",
    "cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "1ba2a248-1a40-4978-8b1e-6a7110318f40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>length</th>\n",
       "      <th>offset</th>\n",
       "      <th>linebases</th>\n",
       "      <th>linewidth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>89168492</td>\n",
       "      <td>82</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>93306550</td>\n",
       "      <td>90283262</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>47367863</td>\n",
       "      <td>184756225</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067198.1</td>\n",
       "      <td>48328128</td>\n",
       "      <td>232716268</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>46771736</td>\n",
       "      <td>281648579</td>\n",
       "      <td>80</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sequence_name    length     offset  linebases  linewidth\n",
       "0   NC_067195.1  89168492         82         80         81\n",
       "1   NC_067196.1  93306550   90283262         80         81\n",
       "2   NC_067197.1  47367863  184756225         80         81\n",
       "3   NC_067198.1  48328128  232716268         80         81\n",
       "4   NC_067199.1  46771736  281648579         80         81"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "size_df = pd.read_csv(\"../../data/GCF_000699445.3_UoM_Shae.V3_genomic.fna.fai\", sep='\\t', header=None, names=['sequence_name', 'length', 'offset', 'linebases', 'linewidth'])\n",
    "size_dict = dict(zip(size_df['sequence_name'], size_df['length']))\n",
    "size_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "398a0e1d-7d32-47c5-a08c-bc35bc0a012a",
   "metadata": {},
   "outputs": [],
   "source": [
    "size_dict = dict(zip(size_df['sequence_name'], size_df['length']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "10ab0633-e76a-4bf9-ab3a-46c1c99eafda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>cluster</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>med_norm_xpehh</th>\n",
       "      <th>len</th>\n",
       "      <th>sig_snvs</th>\n",
       "      <th>bp_to_end</th>\n",
       "      <th>perc_to_end</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>1</td>\n",
       "      <td>85440</td>\n",
       "      <td>86575</td>\n",
       "      <td>6.168670</td>\n",
       "      <td>1135</td>\n",
       "      <td>14</td>\n",
       "      <td>85440</td>\n",
       "      <td>0.000958</td>\n",
       "      <td>NC_067195.1:85440-86575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>2</td>\n",
       "      <td>2450744</td>\n",
       "      <td>2702312</td>\n",
       "      <td>6.369370</td>\n",
       "      <td>251568</td>\n",
       "      <td>35</td>\n",
       "      <td>2450744</td>\n",
       "      <td>0.027484</td>\n",
       "      <td>NC_067195.1:2450744-2702312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>3</td>\n",
       "      <td>15451924</td>\n",
       "      <td>15451973</td>\n",
       "      <td>6.432380</td>\n",
       "      <td>49</td>\n",
       "      <td>9</td>\n",
       "      <td>15451924</td>\n",
       "      <td>0.173289</td>\n",
       "      <td>NC_067195.1:15451924-15451973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>4</td>\n",
       "      <td>64303350</td>\n",
       "      <td>64481575</td>\n",
       "      <td>6.192720</td>\n",
       "      <td>178225</td>\n",
       "      <td>35</td>\n",
       "      <td>24686917</td>\n",
       "      <td>0.276857</td>\n",
       "      <td>NC_067195.1:64303350-64481575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>5</td>\n",
       "      <td>67218409</td>\n",
       "      <td>67220075</td>\n",
       "      <td>-6.708680</td>\n",
       "      <td>1666</td>\n",
       "      <td>129</td>\n",
       "      <td>21948417</td>\n",
       "      <td>0.246145</td>\n",
       "      <td>NC_067195.1:67218409-67220075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>6</td>\n",
       "      <td>69543706</td>\n",
       "      <td>70097099</td>\n",
       "      <td>-6.810760</td>\n",
       "      <td>553393</td>\n",
       "      <td>62</td>\n",
       "      <td>19071393</td>\n",
       "      <td>0.213880</td>\n",
       "      <td>NC_067195.1:69543706-70097099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_067195.1</td>\n",
       "      <td>7</td>\n",
       "      <td>72947731</td>\n",
       "      <td>72947748</td>\n",
       "      <td>6.159410</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>16220744</td>\n",
       "      <td>0.181911</td>\n",
       "      <td>NC_067195.1:72947731-72947748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>8</td>\n",
       "      <td>335206</td>\n",
       "      <td>335218</td>\n",
       "      <td>6.191205</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>335206</td>\n",
       "      <td>0.003593</td>\n",
       "      <td>NC_067196.1:335206-335218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_067196.1</td>\n",
       "      <td>9</td>\n",
       "      <td>92138959</td>\n",
       "      <td>92139077</td>\n",
       "      <td>6.151370</td>\n",
       "      <td>118</td>\n",
       "      <td>32</td>\n",
       "      <td>1167473</td>\n",
       "      <td>0.012512</td>\n",
       "      <td>NC_067196.1:92138959-92139077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>10</td>\n",
       "      <td>1449094</td>\n",
       "      <td>1449134</td>\n",
       "      <td>6.114800</td>\n",
       "      <td>40</td>\n",
       "      <td>3</td>\n",
       "      <td>1449094</td>\n",
       "      <td>0.030592</td>\n",
       "      <td>NC_067197.1:1449094-1449134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>11</td>\n",
       "      <td>16497345</td>\n",
       "      <td>16497348</td>\n",
       "      <td>6.326160</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>16497345</td>\n",
       "      <td>0.348281</td>\n",
       "      <td>NC_067197.1:16497345-16497348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_067197.1</td>\n",
       "      <td>12</td>\n",
       "      <td>46292383</td>\n",
       "      <td>47002311</td>\n",
       "      <td>6.099690</td>\n",
       "      <td>709928</td>\n",
       "      <td>8</td>\n",
       "      <td>365552</td>\n",
       "      <td>0.007717</td>\n",
       "      <td>NC_067197.1:46292383-47002311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_067198.1</td>\n",
       "      <td>13</td>\n",
       "      <td>34371482</td>\n",
       "      <td>34372132</td>\n",
       "      <td>6.151620</td>\n",
       "      <td>650</td>\n",
       "      <td>19</td>\n",
       "      <td>13955996</td>\n",
       "      <td>0.288776</td>\n",
       "      <td>NC_067198.1:34371482-34372132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>14</td>\n",
       "      <td>28684454</td>\n",
       "      <td>28684894</td>\n",
       "      <td>6.791500</td>\n",
       "      <td>440</td>\n",
       "      <td>61</td>\n",
       "      <td>18086842</td>\n",
       "      <td>0.386705</td>\n",
       "      <td>NC_067199.1:28684454-28684894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_067199.1</td>\n",
       "      <td>15</td>\n",
       "      <td>46492141</td>\n",
       "      <td>46551457</td>\n",
       "      <td>6.318750</td>\n",
       "      <td>59316</td>\n",
       "      <td>217</td>\n",
       "      <td>220279</td>\n",
       "      <td>0.004710</td>\n",
       "      <td>NC_067199.1:46492141-46551457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>16</td>\n",
       "      <td>503560</td>\n",
       "      <td>703283</td>\n",
       "      <td>6.174620</td>\n",
       "      <td>199723</td>\n",
       "      <td>13</td>\n",
       "      <td>503560</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>NC_067200.1:503560-703283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>17</td>\n",
       "      <td>2751101</td>\n",
       "      <td>2751781</td>\n",
       "      <td>-6.243965</td>\n",
       "      <td>680</td>\n",
       "      <td>114</td>\n",
       "      <td>2751101</td>\n",
       "      <td>0.124211</td>\n",
       "      <td>NC_067200.1:2751101-2751781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>18</td>\n",
       "      <td>6627231</td>\n",
       "      <td>6627311</td>\n",
       "      <td>-6.259730</td>\n",
       "      <td>80</td>\n",
       "      <td>13</td>\n",
       "      <td>6627231</td>\n",
       "      <td>0.299216</td>\n",
       "      <td>NC_067200.1:6627231-6627311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_067200.1</td>\n",
       "      <td>19</td>\n",
       "      <td>9584838</td>\n",
       "      <td>10269525</td>\n",
       "      <td>-6.231310</td>\n",
       "      <td>684687</td>\n",
       "      <td>212</td>\n",
       "      <td>9584838</td>\n",
       "      <td>0.432750</td>\n",
       "      <td>NC_067200.1:9584838-10269525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_067202.1</td>\n",
       "      <td>20</td>\n",
       "      <td>18613602</td>\n",
       "      <td>18770709</td>\n",
       "      <td>6.239235</td>\n",
       "      <td>157107</td>\n",
       "      <td>10</td>\n",
       "      <td>710480</td>\n",
       "      <td>0.036470</td>\n",
       "      <td>NC_067202.1:18613602-18770709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            chr  cluster     start      stop  med_norm_xpehh     len  \\\n",
       "0   NC_067195.1        1     85440     86575        6.168670    1135   \n",
       "1   NC_067195.1        2   2450744   2702312        6.369370  251568   \n",
       "2   NC_067195.1        3  15451924  15451973        6.432380      49   \n",
       "3   NC_067195.1        4  64303350  64481575        6.192720  178225   \n",
       "4   NC_067195.1        5  67218409  67220075       -6.708680    1666   \n",
       "5   NC_067195.1        6  69543706  70097099       -6.810760  553393   \n",
       "6   NC_067195.1        7  72947731  72947748        6.159410      17   \n",
       "7   NC_067196.1        8    335206    335218        6.191205      12   \n",
       "8   NC_067196.1        9  92138959  92139077        6.151370     118   \n",
       "9   NC_067197.1       10   1449094   1449134        6.114800      40   \n",
       "10  NC_067197.1       11  16497345  16497348        6.326160       3   \n",
       "11  NC_067197.1       12  46292383  47002311        6.099690  709928   \n",
       "12  NC_067198.1       13  34371482  34372132        6.151620     650   \n",
       "13  NC_067199.1       14  28684454  28684894        6.791500     440   \n",
       "14  NC_067199.1       15  46492141  46551457        6.318750   59316   \n",
       "15  NC_067200.1       16    503560    703283        6.174620  199723   \n",
       "16  NC_067200.1       17   2751101   2751781       -6.243965     680   \n",
       "17  NC_067200.1       18   6627231   6627311       -6.259730      80   \n",
       "18  NC_067200.1       19   9584838  10269525       -6.231310  684687   \n",
       "19  NC_067202.1       20  18613602  18770709        6.239235  157107   \n",
       "\n",
       "    sig_snvs  bp_to_end  perc_to_end                       location  \n",
       "0         14      85440     0.000958        NC_067195.1:85440-86575  \n",
       "1         35    2450744     0.027484    NC_067195.1:2450744-2702312  \n",
       "2          9   15451924     0.173289  NC_067195.1:15451924-15451973  \n",
       "3         35   24686917     0.276857  NC_067195.1:64303350-64481575  \n",
       "4        129   21948417     0.246145  NC_067195.1:67218409-67220075  \n",
       "5         62   19071393     0.213880  NC_067195.1:69543706-70097099  \n",
       "6          2   16220744     0.181911  NC_067195.1:72947731-72947748  \n",
       "7          6     335206     0.003593      NC_067196.1:335206-335218  \n",
       "8         32    1167473     0.012512  NC_067196.1:92138959-92139077  \n",
       "9          3    1449094     0.030592    NC_067197.1:1449094-1449134  \n",
       "10         2   16497345     0.348281  NC_067197.1:16497345-16497348  \n",
       "11         8     365552     0.007717  NC_067197.1:46292383-47002311  \n",
       "12        19   13955996     0.288776  NC_067198.1:34371482-34372132  \n",
       "13        61   18086842     0.386705  NC_067199.1:28684454-28684894  \n",
       "14       217     220279     0.004710  NC_067199.1:46492141-46551457  \n",
       "15        13     503560     0.022735      NC_067200.1:503560-703283  \n",
       "16       114    2751101     0.124211    NC_067200.1:2751101-2751781  \n",
       "17        13    6627231     0.299216    NC_067200.1:6627231-6627311  \n",
       "18       212    9584838     0.432750   NC_067200.1:9584838-10269525  \n",
       "19        10     710480     0.036470  NC_067202.1:18613602-18770709  "
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bps=[]\n",
    "percs=[]\n",
    "locations=[]\n",
    "\n",
    "for index, row in cluster_df.iterrows():\n",
    "\n",
    "    chrom = row[\"chr\"]\n",
    "    start = row[\"start\"]\n",
    "    stop = row[\"stop\"]\n",
    "    chrom_size = size_dict[chrom]\n",
    "    \n",
    "    locations.append(f\"{chrom}:{start}-{stop}\")\n",
    "\n",
    "\n",
    "    bp = min([start, chrom_size-stop])\n",
    "    perc = bp/chrom_size\n",
    "\n",
    "    bps.append(bp)\n",
    "    percs.append(perc)\n",
    "\n",
    "\n",
    "cluster_df[\"bp_to_end\"] = bps\n",
    "cluster_df[\"perc_to_end\"]=percs\n",
    "cluster_df[\"location\"] = locations\n",
    "\n",
    "\n",
    "cluster_df.to_csv(\"xpehh_outlier_windows.csv\")\n",
    "cluster_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scan_selection",
   "language": "python",
   "name": "scan_selection"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
